{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0c7a071e-40bb-43fb-a025-e545273ca1b7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Antimicrobial Resistance Knowledge Graph "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b908eb3-11c2-431b-8ecb-7422d2f22e85",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Abstact\n",
    "\n",
    "This study explores the integration of antibiotic resistance data from GenBank and RefSeq databases into the publicly accessible Knowledge Graph, Wikidata. Antibiotic resistance poses a significant global health threat, with the misuse of antibiotics leading to the emergence of resistant strains. The project focuses on modeling proteins, genes, chromosomes, bacterial strains, and species in Wikidata, using a comprehensive dataset from the National Center for Biotechnology Information (NCBI). The data cleaning process involves addressing discrepancies and extracting crucial information from the linked databases. Despite challenges, the implementation in Wikidata progresses, with ongoing efforts to link bacterial strains, genes, and proteins. The study highlights the need for standardized entries in databases and emphasizes the potential impact of integrating antibiotic resistance information into Wikidata for global accessibility and collaborative contributions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cec738cc-9252-4c73-b7c3-78f30e540be5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Background and Motivation\n",
    "\n",
    "Antibiotics play a crucial role in the treatment of bacterial infections. The use of antibiotics annually saves millions of lives, but it also accelerates the growth of livestock in factory farming. Due to the increased and sometimes unnecessary use of antibiotics, individual microorganisms develop resistance, rendering antibiotics ineffective. The development of antibiotic resistance is caused by genetic mutations that occur randomly in bacteria. If these mutations have a positive impact on the bacterium's survival, it survives the antibiotic treatment and passes on the positive traits to other bacteria or the next generation.\n",
    "\n",
    "Improper use of antibiotics, especially in developing countries, has led to the prevalence of many antibiotic-resistant bacteria today. Just as different antibiotics have different mechanisms of action (for example, Beta-Lactam targets the bacterial cell wall), bacteria also develop various defense mechanisms. For instance, bacteria can form an efflux pump, which expels already entered antibiotics from the cell.\n",
    "\n",
    "Infection with antibiotic-resistant bacteria poses a high risk to humans, as medical treatment becomes impossible. The World Health Organization (WHO) classifies Antimicrobial Resistance (AMR) as one of the three greatest medical threats, referring to it as a silent pandemic due to the high annual death toll (1.27 million people). By 2050, the WHO estimates that the number could rise to 10 million deaths per year, far surpassing the annual mortality rate of cancer.\n",
    "\n",
    "Although various health organizations are aware of numerous bacterial strains encoding antibiotic-resistant proteins, these databases are not interconnected, despite the urgent need for such integration. This work focuses on incorporating the largest publicly accessible database from the National Center for Biotechnology Information (NCBI) into a publicly accessible Knowledge Graph (Wikidata). Medial staff, scientists and interested people can easily access the knowledge from Wikidata. [1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe45f883-19a1-488f-a027-0240be90fe70",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Material and Methods\n",
    "\n",
    "Since the data in the publicly accessible Wikidata Knowledge Graph is intended to be available to everyone, the implementation concept follows the modeling of proteins, genes, chromosomes, bacterial strains, and bacterial species. The premise is that the bacterial species is already present in Wikidata, and only the further sequence needs to be modeled.\n",
    "\n",
    "As a data source, a public database from the National Center for Biotechnology Information (NIH) is chosen [2]. This database consists of approximately 10,000 proteins that cause antibiotic resistances in bacteria. The protein is linked to the type of antibiotic resistance, as well as to links to nucleotides and proteins in the Refseq and GenBank databases. To model the sequence described above, consisting of protein, gene, chromosome, and bacterial strain starting from the bacterial species, the names of the gene and bacterial strain must be extracted from the linked Refseq and GenBank databases. Using the \"bioservice EUtils\" and the Refseq or GenBank references, the aforementioned information is extracted from the databases. Redundancies are deliberately built (e.g., extracting the bacterial species via Refseq and GenBank, as well as via protein and nucleotide) to subsequently combine information or choose the most complete data source.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e7a1f49c-dbf1-4ce6-9104-c54b91f5ed17",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import urllib\n",
    "import time \n",
    "import numpy as np \n",
    "import bioservice_fetcher as biof \n",
    "import os \n",
    "from typing import Optional "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e328c2b-56f0-43b8-9211-3856cd73464e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Fetch data from NCBI database \n",
    "\n",
    "The read dataframe contains keys to access the protein, nucleotide either via Reference Sequence (RefSeq) database or genbank. Lets load the first two columns of the dataframe and read some values. First try is to get the data from the RefSeq-database.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4f4fe93-91d3-48bb-9a65-626aea9a69cf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele</th>\n",
       "      <th>gene_family</th>\n",
       "      <th>whitelisted_taxa</th>\n",
       "      <th>product_name</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>refseq_protein_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>refseq_gene</th>\n",
       "      <th>refseq_protein2</th>\n",
       "      <th>refseq_genome</th>\n",
       "      <th>refseq_organism</th>\n",
       "      <th>refseq_tax_id</th>\n",
       "      <th>genbank_organism</th>\n",
       "      <th>genbank_strain</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_tax_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3774</th>\n",
       "      <td>blaOXA-823</td>\n",
       "      <td>blaOXA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OXA-10 family class D beta-lactamase OXA-823</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>WP_136512103.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Pseudomonas aeruginosa HUPM19015969 blaOXA</td>\n",
       "      <td>OXA-10 family class D beta-lactamase OXA-823</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>taxon:287</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>HUPM19015969</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>HUPM19015969</td>\n",
       "      <td>taxon:287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6706</th>\n",
       "      <td>NaN</td>\n",
       "      <td>narB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ionophore ABC transporter permease subunit NarB</td>\n",
       "      <td>plus</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>IONOPHORE</td>\n",
       "      <td>MADURAMICIN/NARASIN/SALINOMYCIN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>WT1145</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>WT1145</td>\n",
       "      <td>taxon:1352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>aac(6')-29a</td>\n",
       "      <td>aac(6')-29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside N-acetyltransferase AAC(6')-29a</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>WP_064190968.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Pseudomonas aeruginosa aac(6')-29</td>\n",
       "      <td>aminoglycoside N-acetyltransferase AAC(6')-29a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>taxon:287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>taxon:287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7508</th>\n",
       "      <td>NaN</td>\n",
       "      <td>tet(D)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tetracycline efflux MFS transporter Tet(D)</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>TETRACYCLINE</td>\n",
       "      <td>TETRACYCLINE</td>\n",
       "      <td>WP_001039466.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Shigella sonnei 119 tet(D)</td>\n",
       "      <td>tetracycline efflux MFS transporter Tet(D)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shigella sonnei</td>\n",
       "      <td>taxon:624</td>\n",
       "      <td>Shigella sonnei</td>\n",
       "      <td>119</td>\n",
       "      <td>Shigella sonnei</td>\n",
       "      <td>119</td>\n",
       "      <td>taxon:624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1398</th>\n",
       "      <td>blaCAR-1</td>\n",
       "      <td>blaCAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subclass B3 metallo-beta-lactamase CAR-1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CARBAPENEM</td>\n",
       "      <td>WP_011094382.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Pectobacterium atrosepticum SCRI1043 blaCAR</td>\n",
       "      <td>subclass B3 metallo-beta-lactamase CAR-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pectobacterium atrosepticum SCRI1043</td>\n",
       "      <td>taxon:218491</td>\n",
       "      <td>Pectobacterium atrosepticum SCRI1043</td>\n",
       "      <td>SCRI1043</td>\n",
       "      <td>Pectobacterium atrosepticum SCRI1043</td>\n",
       "      <td>SCRI1043</td>\n",
       "      <td>taxon:218491</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           allele gene_family whitelisted_taxa  \\\n",
       "3774   blaOXA-823      blaOXA              NaN   \n",
       "6706          NaN        narB              NaN   \n",
       "139   aac(6')-29a  aac(6')-29              NaN   \n",
       "7508          NaN      tet(D)              NaN   \n",
       "1398     blaCAR-1      blaCAR              NaN   \n",
       "\n",
       "                                         product_name scope type subtype  \\\n",
       "3774     OXA-10 family class D beta-lactamase OXA-823  core  AMR     AMR   \n",
       "6706  ionophore ABC transporter permease subunit NarB  plus  AMR     AMR   \n",
       "139    aminoglycoside N-acetyltransferase AAC(6')-29a  core  AMR     AMR   \n",
       "7508       tetracycline efflux MFS transporter Tet(D)  core  AMR     AMR   \n",
       "1398         subclass B3 metallo-beta-lactamase CAR-1  core  AMR     AMR   \n",
       "\n",
       "               class                         subclass  \\\n",
       "3774     BETA-LACTAM                      BETA-LACTAM   \n",
       "6706       IONOPHORE  MADURAMICIN/NARASIN/SALINOMYCIN   \n",
       "139   AMINOGLYCOSIDE                   AMINOGLYCOSIDE   \n",
       "7508    TETRACYCLINE                     TETRACYCLINE   \n",
       "1398     BETA-LACTAM                       CARBAPENEM   \n",
       "\n",
       "     refseq_protein_accession  ...  \\\n",
       "3774           WP_136512103.1  ...   \n",
       "6706                      NaN  ...   \n",
       "139            WP_064190968.1  ...   \n",
       "7508           WP_001039466.1  ...   \n",
       "1398           WP_011094382.1  ...   \n",
       "\n",
       "                                      refseq_gene  \\\n",
       "3774   Pseudomonas aeruginosa HUPM19015969 blaOXA   \n",
       "6706                                          NaN   \n",
       "139             Pseudomonas aeruginosa aac(6')-29   \n",
       "7508                   Shigella sonnei 119 tet(D)   \n",
       "1398  Pectobacterium atrosepticum SCRI1043 blaCAR   \n",
       "\n",
       "                                     refseq_protein2 refseq_genome  \\\n",
       "3774    OXA-10 family class D beta-lactamase OXA-823           NaN   \n",
       "6706                                             NaN           NaN   \n",
       "139   aminoglycoside N-acetyltransferase AAC(6')-29a           NaN   \n",
       "7508      tetracycline efflux MFS transporter Tet(D)           NaN   \n",
       "1398        subclass B3 metallo-beta-lactamase CAR-1           NaN   \n",
       "\n",
       "                           refseq_organism refseq_tax_id  \\\n",
       "3774                Pseudomonas aeruginosa     taxon:287   \n",
       "6706                                   NaN           NaN   \n",
       "139                 Pseudomonas aeruginosa     taxon:287   \n",
       "7508                       Shigella sonnei     taxon:624   \n",
       "1398  Pectobacterium atrosepticum SCRI1043  taxon:218491   \n",
       "\n",
       "                          genbank_organism  genbank_strain  \\\n",
       "3774                Pseudomonas aeruginosa    HUPM19015969   \n",
       "6706                  Enterococcus faecium          WT1145   \n",
       "139                                    NaN             NaN   \n",
       "7508                       Shigella sonnei             119   \n",
       "1398  Pectobacterium atrosepticum SCRI1043        SCRI1043   \n",
       "\n",
       "                         genbank_organism2  genbank_strain2  genbank_tax_id  \n",
       "3774                Pseudomonas aeruginosa     HUPM19015969       taxon:287  \n",
       "6706                  Enterococcus faecium           WT1145      taxon:1352  \n",
       "139                 Pseudomonas aeruginosa              NaN       taxon:287  \n",
       "7508                       Shigella sonnei              119       taxon:624  \n",
       "1398  Pectobacterium atrosepticum SCRI1043         SCRI1043    taxon:218491  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reads everything from the linked refseq and genbank databases that could possible be interesting for this project\n",
    "\n",
    "def fetch_data(read_from_web: bool = False, amount: Optional[int] = None) -> pd.DataFrame: \n",
    "    \"\"\"\n",
    "    if fetch_data_switch is True data will be fetched from NCBI and interesting things will be read from genbank or refseq database -- caution: Takes for ages \n",
    "    else data will be read from last time -- should be used in most of the cases \n",
    "    \"\"\"\n",
    "    if read_from_web: \n",
    "        # will take ~ 12h \n",
    "        url = \"https://ftp.ncbi.nlm.nih.gov/pathogen/Antimicrobial_resistance/AMRFinderPlus/database/latest/ReferenceGeneCatalog.txt\"\n",
    "        df = pd.read_csv(urllib.request.urlopen(url), delimiter=\"\\t\")\n",
    "        if amount is not None:\n",
    "            df = df.sample(amount)\n",
    "        df[[\"refseq_parent_taxon\", \"refseq_protein\", \"refseq_parent_taxon2\"]] = df.apply(biof.get_protein_and_parent, axis=1)\n",
    "        print(1)\n",
    "        df[[\"refseq_gene\", \"refseq_protein2\", \"refseq_genome\", \"refseq_organism\", \"refseq_tax_id\"]] = df.apply(biof.get_strain_and_gene, axis=1)\n",
    "        print(2)\n",
    "        df[[\"genbank_organism\", \"genbank_strain\"]] = df.apply(biof.get_organism_strain_via_prot,axis=1)\n",
    "        print(3)\n",
    "        df[[\"genbank_organism2\", \"genbank_strain2\", \"genbank_tax_id\"]] = df.apply(biof.get_organism_strain_via_nuc, axis=1)\n",
    "        df.to_csv(\"resistance_df.csv\", index=False)\n",
    "    else: \n",
    "        if not os.path.exists(\"resistance_df.csv\"): \n",
    "            print(\"Cannot read from hard drive because file does not exist -- set read from web switch to True\")\n",
    "            return None\n",
    "        df = pd.read_csv(\"resistance_df.csv\")\n",
    "    return df\n",
    "\n",
    "\n",
    "df = fetch_data(False)\n",
    "df.sample(5, random_state=19)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffd50fed-0478-4c5e-86f5-77b4d268ea51",
   "metadata": {
    "tags": []
   },
   "source": [
    "## STRAIN\n",
    "\n",
    "Because there is quite a bit of data present, my goal is to extract the right name of the bacterial strain. Therefore, I am going to find the best data source of organism name and combine it with the best source of the exact strain name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "16a7848f-4bdd-4011-837e-dcd9f3c1adf0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['allele', 'gene_family', 'whitelisted_taxa', 'product_name', 'scope',\n",
       "       'type', 'subtype', 'class', 'subclass', 'refseq_protein_accession',\n",
       "       'refseq_nucleotide_accession', 'curated_refseq_start',\n",
       "       'genbank_protein_accession', 'genbank_nucleotide_accession',\n",
       "       'genbank_strand', 'genbank_start', 'genbank_stop', 'refseq_strand',\n",
       "       'refseq_start', 'refseq_stop', 'pubmed_reference', 'blacklisted_taxa',\n",
       "       'synonyms', 'hierarchy_node', 'db_version', 'refseq_parent_taxon',\n",
       "       'refseq_protein', 'refseq_parent_taxon2', 'refseq_gene',\n",
       "       'refseq_protein2', 'refseq_genome', 'refseq_organism', 'refseq_tax_id',\n",
       "       'genbank_organism', 'genbank_strain', 'genbank_organism2',\n",
       "       'genbank_strain2', 'genbank_tax_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e2217a5c-536c-46c0-ad9b-8aa04f75f4de",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq_parent_taxon</th>\n",
       "      <th>refseq_parent_taxon2</th>\n",
       "      <th>refseq_organism</th>\n",
       "      <th>genbank_organism</th>\n",
       "      <th>genbank_organism2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acidiphilium multivorum</td>\n",
       "      <td>Acidiphilium multivorum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9222</th>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shigella sonnei</td>\n",
       "      <td>Shigella sonnei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8816</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3993</th>\n",
       "      <td>Klebsiella oxytoca</td>\n",
       "      <td>Klebsiella oxytoca</td>\n",
       "      <td>Klebsiella oxytoca</td>\n",
       "      <td>Klebsiella oxytoca</td>\n",
       "      <td>Klebsiella oxytoca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4874</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956</th>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Citrobacter freundii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas putida</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas putida</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3978</th>\n",
       "      <td>Klebsiella michiganensis</td>\n",
       "      <td>Klebsiella michiganensis</td>\n",
       "      <td>Klebsiella michiganensis</td>\n",
       "      <td>Klebsiella michiganensis</td>\n",
       "      <td>Klebsiella michiganensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8247</th>\n",
       "      <td>Enterococcus</td>\n",
       "      <td>Enterococcus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2520</th>\n",
       "      <td>Lelliottia amnigena</td>\n",
       "      <td>Lelliottia amnigena</td>\n",
       "      <td>Lelliottia amnigena</td>\n",
       "      <td>Lelliottia amnigena</td>\n",
       "      <td>Lelliottia amnigena</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           refseq_parent_taxon      refseq_parent_taxon2  \\\n",
       "758                        NaN                       NaN   \n",
       "9222        Enterobacteriaceae        Enterobacteriaceae   \n",
       "8816                  Bacteria                  Bacteria   \n",
       "3993        Klebsiella oxytoca        Klebsiella oxytoca   \n",
       "4874          Escherichia coli          Escherichia coli   \n",
       "6956      Citrobacter freundii      Citrobacter freundii   \n",
       "277           Pseudomonadaceae          Pseudomonadaceae   \n",
       "3978  Klebsiella michiganensis  Klebsiella michiganensis   \n",
       "8247              Enterococcus              Enterococcus   \n",
       "2520       Lelliottia amnigena       Lelliottia amnigena   \n",
       "\n",
       "               refseq_organism                           genbank_organism  \\\n",
       "758                        NaN                    Acidiphilium multivorum   \n",
       "9222                       NaN                            Shigella sonnei   \n",
       "8816                       NaN  Escherichia coli str. K-12 substr. MG1655   \n",
       "3993        Klebsiella oxytoca                         Klebsiella oxytoca   \n",
       "4874          Escherichia coli                           Escherichia coli   \n",
       "6956      Citrobacter freundii                                        NaN   \n",
       "277         Pseudomonas putida                                        NaN   \n",
       "3978  Klebsiella michiganensis                   Klebsiella michiganensis   \n",
       "8247                       NaN                       Enterococcus faecium   \n",
       "2520       Lelliottia amnigena                        Lelliottia amnigena   \n",
       "\n",
       "                              genbank_organism2  \n",
       "758                     Acidiphilium multivorum  \n",
       "9222                            Shigella sonnei  \n",
       "8816  Escherichia coli str. K-12 substr. MG1655  \n",
       "3993                         Klebsiella oxytoca  \n",
       "4874                           Escherichia coli  \n",
       "6956                       Citrobacter freundii  \n",
       "277                          Pseudomonas putida  \n",
       "3978                   Klebsiella michiganensis  \n",
       "8247                       Enterococcus faecium  \n",
       "2520                        Lelliottia amnigena  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Just by looking into this small random sampled dataframe \"refseq organism\", \"genbank_organsim\" and \"genbank_organism2\" yield similar results, \n",
    "# altough \"refseq_organism\" has less information\n",
    "# \"refseq_parent_taxon\" and \"refseq_parent_taxon2\" sometimes have the same information as the others and sometimes have a much higher taxon (e.g. line 8816 \"Bacteria\")\n",
    "# For extracting the right organism name I am going to look closer into \"refseq_organism\", \"genbank_organism\" and \"ganbank_organism2\"\n",
    "df[[\"refseq_parent_taxon\", \"refseq_parent_taxon2\",  \"refseq_organism\", \"genbank_organism\", \"genbank_organism2\"]].sample(10, random_state=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3389baea-04e6-4f2b-a1db-69005a380fa8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genbank_organism      368\n",
       "genbank_organism2       0\n",
       "refseq_organism      2499\n",
       "dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The above impression solifies: refseq organism obivously contains the most empty fields\n",
    "df[[\"genbank_organism\", \"genbank_organism2\", \"refseq_organism\"]].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1ed7e512-a2db-41b6-b712-06ae8a038718",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organism</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>refseq_organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7123</th>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6122</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>Enterococcus sp. JM4C</td>\n",
       "      <td>Enterococcus sp. JM4C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8458</th>\n",
       "      <td>Campylobacter jejuni subsp. jejuni NCTC 11168 ...</td>\n",
       "      <td>Campylobacter jejuni subsp. jejuni NCTC 11168 ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6540</th>\n",
       "      <td>Enterobacter roggenkampii</td>\n",
       "      <td>Enterobacter roggenkampii</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Acinetobacter baumannii AB4A3</td>\n",
       "      <td>Acinetobacter baumannii AB4A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8553</th>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5718</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3126</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5948</th>\n",
       "      <td>Bacillus bingmayongensis</td>\n",
       "      <td>Bacillus bingmayongensis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       genbank_organism  \\\n",
       "7123                              Staphylococcus aureus   \n",
       "6122                                   Escherichia coli   \n",
       "770                               Enterococcus sp. JM4C   \n",
       "8458  Campylobacter jejuni subsp. jejuni NCTC 11168 ...   \n",
       "6540                          Enterobacter roggenkampii   \n",
       "999                                                 NaN   \n",
       "8553                             Pseudomonas aeruginosa   \n",
       "5718                                   Escherichia coli   \n",
       "3126                                                NaN   \n",
       "5948                           Bacillus bingmayongensis   \n",
       "\n",
       "                                      genbank_organism2  \\\n",
       "7123                              Staphylococcus aureus   \n",
       "6122                                   Escherichia coli   \n",
       "770                               Enterococcus sp. JM4C   \n",
       "8458  Campylobacter jejuni subsp. jejuni NCTC 11168 ...   \n",
       "6540                          Enterobacter roggenkampii   \n",
       "999                       Acinetobacter baumannii AB4A3   \n",
       "8553                             Pseudomonas aeruginosa   \n",
       "5718                                   Escherichia coli   \n",
       "3126                             Pseudomonas aeruginosa   \n",
       "5948                           Bacillus bingmayongensis   \n",
       "\n",
       "                    refseq_organism  \n",
       "7123                            NaN  \n",
       "6122                            NaN  \n",
       "770                             NaN  \n",
       "8458                            NaN  \n",
       "6540                            NaN  \n",
       "999   Acinetobacter baumannii AB4A3  \n",
       "8553                            NaN  \n",
       "5718                            NaN  \n",
       "3126         Pseudomonas aeruginosa  \n",
       "5948                            NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#cases where organisms in genbank via protein (genbank_organism), genbank via nuclotide and refseq do not match \n",
    "\n",
    "diff_organism_df = df.loc[(df[\"genbank_organism\"] != df[\"genbank_organism2\"]) | (df[\"genbank_organism\"] != df[\"refseq_organism\"]), [\"genbank_organism\", \n",
    "                                                                                                                                    \"genbank_organism2\", \n",
    "                                                                                                                                    \"refseq_organism\"]]\n",
    "diff_organism_df.sample(10, random_state=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "30f6564d-24b1-47ee-ab09-3b437ebafc95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organism</th>\n",
       "      <th>genbank_organism2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5475</th>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5697</th>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5721</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5942</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6214</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6217</th>\n",
       "      <td>Staphylococcus epidermidis RP62A</td>\n",
       "      <td>Staphylococcus epidermidis RP62A phage SP-beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6220</th>\n",
       "      <td>Bacillus phage lambda Ba03</td>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6257</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6258</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6263</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6264</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6265</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6275</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6305</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6306</th>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6542</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6543</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6724</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6869</th>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6870</th>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7106</th>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7168</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7176</th>\n",
       "      <td>Phage Gifsy-2</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7178</th>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7417</th>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7418</th>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7419</th>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7420</th>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7421</th>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8027</th>\n",
       "      <td>Streptococcus pneumoniae AP200</td>\n",
       "      <td>Streptococcus phage PhiSpn_200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8036</th>\n",
       "      <td>Streptococcus pneumoniae AP200</td>\n",
       "      <td>Streptococcus phage PhiSpn_200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8042</th>\n",
       "      <td>Streptococcus pneumoniae AP200</td>\n",
       "      <td>Streptococcus phage PhiSpn_200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9308</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       genbank_organism  \\\n",
       "126   Salmonella enterica subsp. enterica serovar Ty...   \n",
       "601                        Bacillus anthracis str. Ames   \n",
       "5475                              Escherichia coli K-12   \n",
       "5697                       Clostridioides difficile 630   \n",
       "5721                         Bacillus cereus ATCC 14579   \n",
       "5942                         Bacillus cereus ATCC 14579   \n",
       "6214                         Bacillus cereus ATCC 14579   \n",
       "6217                   Staphylococcus epidermidis RP62A   \n",
       "6220                         Bacillus phage lambda Ba03   \n",
       "6257  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "6258  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "6263                         Bacillus cereus ATCC 14579   \n",
       "6264                         Bacillus cereus ATCC 14579   \n",
       "6265                         Bacillus cereus ATCC 14579   \n",
       "6275                         Bacillus cereus ATCC 14579   \n",
       "6305                         Bacillus cereus ATCC 14579   \n",
       "6306                       Bacillus anthracis str. Ames   \n",
       "6542  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "6543  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "6724                         Bacillus cereus ATCC 14579   \n",
       "6869                         Bacillus cereus ATCC 14579   \n",
       "6870                       Bacillus anthracis str. Ames   \n",
       "7106                       Bacillus anthracis str. Ames   \n",
       "7168  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "7176                                      Phage Gifsy-2   \n",
       "7178                       Bacillus anthracis str. Ames   \n",
       "7417                       Clostridioides difficile 630   \n",
       "7418                       Clostridioides difficile 630   \n",
       "7419                       Clostridioides difficile 630   \n",
       "7420                       Clostridioides difficile 630   \n",
       "7421                       Clostridioides difficile 630   \n",
       "8027                     Streptococcus pneumoniae AP200   \n",
       "8036                     Streptococcus pneumoniae AP200   \n",
       "8042                     Streptococcus pneumoniae AP200   \n",
       "9308  Salmonella enterica subsp. enterica serovar Ty...   \n",
       "\n",
       "                                   genbank_organism2  \n",
       "126                           Salmonella virus Fels2  \n",
       "601                       Bacillus phage lambda Ba02  \n",
       "5475                                Escherichia coli  \n",
       "5697                   Peptoclostridium phage p630P2  \n",
       "5721                         Bacillus phage phBC6A52  \n",
       "5942                         Bacillus phage phBC6A52  \n",
       "6214                         Bacillus phage phBC6A52  \n",
       "6217  Staphylococcus epidermidis RP62A phage SP-beta  \n",
       "6220                      Bacillus phage lambda Ba02  \n",
       "6257                          Salmonella virus Fels2  \n",
       "6258                          Salmonella virus Fels2  \n",
       "6263                         Bacillus phage phBC6A52  \n",
       "6264                         Bacillus phage phBC6A52  \n",
       "6265                         Bacillus phage phBC6A52  \n",
       "6275                         Bacillus phage phBC6A52  \n",
       "6305                         Bacillus phage phBC6A52  \n",
       "6306                      Bacillus phage lambda Ba02  \n",
       "6542                          Salmonella virus Fels2  \n",
       "6543                          Salmonella virus Fels2  \n",
       "6724                         Bacillus phage phBC6A52  \n",
       "6869                         Bacillus phage phBC6A52  \n",
       "6870                      Bacillus phage lambda Ba02  \n",
       "7106                      Bacillus phage lambda Ba02  \n",
       "7168                          Salmonella virus Fels2  \n",
       "7176                          Salmonella virus Fels2  \n",
       "7178                      Bacillus phage lambda Ba02  \n",
       "7417                   Peptoclostridium phage p630P2  \n",
       "7418                   Peptoclostridium phage p630P2  \n",
       "7419                   Peptoclostridium phage p630P2  \n",
       "7420                   Peptoclostridium phage p630P2  \n",
       "7421                   Peptoclostridium phage p630P2  \n",
       "8027                  Streptococcus phage PhiSpn_200  \n",
       "8036                  Streptococcus phage PhiSpn_200  \n",
       "8042                  Streptococcus phage PhiSpn_200  \n",
       "9308                          Salmonella virus Fels2  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# As can be seen there are 35 issues where organism found via genbank (nucleotid) and via genbank (protein) are different \n",
    "# What should be done here? Are these alternative names? -- Expert knowledge requiered: I'm going to drop them \n",
    "(x1 := diff_organism_df.loc[diff_organism_df[\"genbank_organism\"] != diff_organism_df[\"genbank_organism2\"], [\"genbank_organism\", \"genbank_organism2\"]].dropna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "58ac0927-c085-4a51-b489-c5895b707fcc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organism</th>\n",
       "      <th>refseq_organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2032</th>\n",
       "      <td>Acinetobacter sp.</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4045</th>\n",
       "      <td>Klebsiella michiganensis</td>\n",
       "      <td>Klebsiella michiganensis M5al</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5056</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6220</th>\n",
       "      <td>Bacillus phage lambda Ba03</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6236</th>\n",
       "      <td>Cytobacillus massiliigabonensis</td>\n",
       "      <td>Bacillus massiliigabonensis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     genbank_organism  \\\n",
       "2032                Acinetobacter sp.   \n",
       "4045         Klebsiella michiganensis   \n",
       "5056              Salmonella enterica   \n",
       "6220       Bacillus phage lambda Ba03   \n",
       "6236  Cytobacillus massiliigabonensis   \n",
       "\n",
       "                                        refseq_organism  \n",
       "2032                            Acinetobacter baumannii  \n",
       "4045                      Klebsiella michiganensis M5al  \n",
       "5056  Salmonella enterica subsp. enterica serovar In...  \n",
       "6220                       Bacillus anthracis str. Ames  \n",
       "6236                        Bacillus massiliigabonensis  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Furthermore there are 5 issues between genbank via nucleotide and refseq \n",
    "# Are these alternative names? Expert knowledge requiered: I'm going to drop them as I'm not sure\n",
    "(x2 := diff_organism_df.loc[diff_organism_df[\"genbank_organism\"] != diff_organism_df[\"refseq_organism\"], [\"genbank_organism\", \n",
    "                                                                                                          \"refseq_organism\"]].dropna())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49011782-010e-46c4-83ab-a835a96c04f7",
   "metadata": {
    "tags": []
   },
   "source": [
    "I am going to choose the organism found by genbank via nucleotid as base organism, because it has no NaN and seems to be complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "74a6241c-8489-47f3-82bc-541618faa6b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# drop data instances where I'm unsure \n",
    "df.drop(set(x1.index).union(set(x2.index)), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1c2c33a6-6342-4947-b37f-d1b62d2ff08e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_strain</th>\n",
       "      <th>genbank_organism2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [genbank_strain2, genbank_strain, genbank_organism2]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Genbank_strain carrys either redundant information or no information at all -- genbank_strain2 will therefore be selected\n",
    "df.loc[(df[\"genbank_strain2\"] != df[\"genbank_strain\"]) & ~(df[\"genbank_strain\"].isna()), [\"genbank_strain2\", \"genbank_strain\", \"genbank_organism2\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6c37ce10-be7a-4743-b8b4-0d93b538e598",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_organism2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9018</th>\n",
       "      <td>FA19</td>\n",
       "      <td>Neisseria gonorrhoeae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7743</th>\n",
       "      <td>MLG080-1</td>\n",
       "      <td>Clostridium sp. MLG080-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9165</th>\n",
       "      <td>WHO_U</td>\n",
       "      <td>Neisseria gonorrhoeae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3013</th>\n",
       "      <td>2318902</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FC1K</td>\n",
       "      <td>Mycolicibacterium fortuitum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6795</th>\n",
       "      <td>SKLX003475</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1363</th>\n",
       "      <td>G4074</td>\n",
       "      <td>Elizabethkingia miricola</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4295</th>\n",
       "      <td>185584</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9014</th>\n",
       "      <td>FA19</td>\n",
       "      <td>Neisseria gonorrhoeae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5168</th>\n",
       "      <td>13S00929-3</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genbank_strain2            genbank_organism2\n",
       "9018            FA19        Neisseria gonorrhoeae\n",
       "7743        MLG080-1     Clostridium sp. MLG080-1\n",
       "9165           WHO_U        Neisseria gonorrhoeae\n",
       "3013         2318902       Pseudomonas aeruginosa\n",
       "4               FC1K  Mycolicibacterium fortuitum\n",
       "6795      SKLX003475        Klebsiella pneumoniae\n",
       "1363           G4074     Elizabethkingia miricola\n",
       "4295          185584       Pseudomonas aeruginosa\n",
       "9014            FA19        Neisseria gonorrhoeae\n",
       "5168      13S00929-3             Escherichia coli"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Often the strain name in as expected to be in genbank_strain2, somtimes it is already included in genbank organism\n",
    "# e.g. line 7743: Clostridium sp. MLG080-1 \n",
    "df.loc[:, [\"genbank_strain2\", \"genbank_organism2\"]].sample(10, random_state=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f96c2ab3-5c47-4b2d-8201-46c019530152",
   "metadata": {
    "tags": []
   },
   "source": [
    "To find the correct and full strain name, I will combine \"genbank_strain2\" and \"genbank_organism2\", only if strain is not included in organism. Else, the organism itself will be selected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4f515de2-4476-4873-a640-d93c2e5c5d9c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5358</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>850</th>\n",
       "      <td>HA-2</td>\n",
       "      <td>Hafnia alvei</td>\n",
       "      <td>Hafnia alvei HA-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9038</th>\n",
       "      <td>VRCO0432</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae VRCO0432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4021</th>\n",
       "      <td>SG271</td>\n",
       "      <td>Klebsiella spallanzanii</td>\n",
       "      <td>Klebsiella spallanzanii SG271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7357</th>\n",
       "      <td>NIPH56</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>Acinetobacter baumannii NIPH56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5862</th>\n",
       "      <td>13H1</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli 13H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529</th>\n",
       "      <td>HD24</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae HD24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>NaN</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8818</th>\n",
       "      <td>K-12</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4861</th>\n",
       "      <td>KK19</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae KK19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genbank_strain2                          genbank_organism2  \\\n",
       "5358             NaN                      Staphylococcus aureus   \n",
       "850             HA-2                               Hafnia alvei   \n",
       "9038        VRCO0432                      Klebsiella pneumoniae   \n",
       "4021           SG271                    Klebsiella spallanzanii   \n",
       "7357          NIPH56                    Acinetobacter baumannii   \n",
       "5862            13H1                           Escherichia coli   \n",
       "2529            HD24                      Klebsiella pneumoniae   \n",
       "613              NaN                       uncultured bacterium   \n",
       "8818            K-12  Escherichia coli str. K-12 substr. MG1655   \n",
       "4861            KK19                      Klebsiella pneumoniae   \n",
       "\n",
       "                                         strain  \n",
       "5358                      Staphylococcus aureus  \n",
       "850                           Hafnia alvei HA-2  \n",
       "9038             Klebsiella pneumoniae VRCO0432  \n",
       "4021              Klebsiella spallanzanii SG271  \n",
       "7357             Acinetobacter baumannii NIPH56  \n",
       "5862                      Escherichia coli 13H1  \n",
       "2529                 Klebsiella pneumoniae HD24  \n",
       "613                        uncultured bacterium  \n",
       "8818  Escherichia coli str. K-12 substr. MG1655  \n",
       "4861                 Klebsiella pneumoniae KK19  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def check_for_combination(df_row: pd.Series) -> bool:\n",
    "    \"\"\"\n",
    "    Checks if genbank_organism and genbank_strain should be connected to one strain\n",
    "    \"\"\"\n",
    "    if not isinstance(df_row[\"genbank_strain2\"], str): \n",
    "        # don't combine organism and strain if is NaN --> Is this correct? \n",
    "        return False\n",
    "    return not (df_row[\"genbank_strain2\"].upper() in df_row[\"genbank_organism2\"].upper())\n",
    "    \n",
    "    \n",
    "df[\"strain\"] = np.where(df.apply(check_for_combination , axis=1), \n",
    "                           df[\"genbank_organism2\"] + \" \" + df[\"genbank_strain2\"], \n",
    "                           df[\"genbank_organism2\"])\n",
    "df[[\"genbank_strain2\", \"genbank_organism2\", \"strain\"]].sample(10, random_state=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fba161c6-845c-4f9f-845d-22ece9ecdba8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6455</th>\n",
       "      <td>EC3769</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli EC3769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5176</th>\n",
       "      <td>CLSiS 1590/96</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae CLSiS 1590/96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8198</th>\n",
       "      <td>VC4477</td>\n",
       "      <td>Burkholderia cenocepacia</td>\n",
       "      <td>Burkholderia cenocepacia VC4477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2566</th>\n",
       "      <td>blaCLHK-3</td>\n",
       "      <td>Laribacter hongkongensis</td>\n",
       "      <td>Laribacter hongkongensis blaCLHK-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466</th>\n",
       "      <td>BS</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli BS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5081</th>\n",
       "      <td>3343</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli 3343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8839</th>\n",
       "      <td>PAO1</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>268/2C</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>Acinetobacter baumannii 268/2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3345</th>\n",
       "      <td>XM1570</td>\n",
       "      <td>Acinetobacter calcoaceticus</td>\n",
       "      <td>Acinetobacter calcoaceticus XM1570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9152</th>\n",
       "      <td>PNUSAS062732</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>Salmonella enterica PNUSAS062732</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genbank_strain2            genbank_organism2  \\\n",
       "6455          EC3769             Escherichia coli   \n",
       "5176   CLSiS 1590/96        Klebsiella pneumoniae   \n",
       "8198          VC4477     Burkholderia cenocepacia   \n",
       "2566       blaCLHK-3     Laribacter hongkongensis   \n",
       "1466              BS             Escherichia coli   \n",
       "5081            3343             Escherichia coli   \n",
       "8839            PAO1  Pseudomonas aeruginosa PAO1   \n",
       "3313          268/2C      Acinetobacter baumannii   \n",
       "3345          XM1570  Acinetobacter calcoaceticus   \n",
       "9152    PNUSAS062732          Salmonella enterica   \n",
       "\n",
       "                                   strain  \n",
       "6455              Escherichia coli EC3769  \n",
       "5176  Klebsiella pneumoniae CLSiS 1590/96  \n",
       "8198      Burkholderia cenocepacia VC4477  \n",
       "2566   Laribacter hongkongensis blaCLHK-3  \n",
       "1466                  Escherichia coli BS  \n",
       "5081                Escherichia coli 3343  \n",
       "8839          Pseudomonas aeruginosa PAO1  \n",
       "3313       Acinetobacter baumannii 268/2C  \n",
       "3345   Acinetobacter calcoaceticus XM1570  \n",
       "9152     Salmonella enterica PNUSAS062732  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# As can be seen sometimes strain now contains only the parent taxon name (e.g. line Staphylococcus aureus), which is not correct. \n",
    "# I will drop instances where strain is no longer than two words. Expert knowledge required  \n",
    "df = df[~(df[\"strain\"].str.split().str.len() <= 2)]\n",
    "df[[\"genbank_strain2\", \"genbank_organism2\", \"strain\"]].sample(10, random_state=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7358438-6e79-4225-88f1-ff9adc7cd87d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "711fdde7-8585-4828-8a74-c386a34ee7de",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_family</th>\n",
       "      <th>refseq_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4850</th>\n",
       "      <td>blaSHV</td>\n",
       "      <td>Klebsiella pneumoniae 90088 blaSHV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3156</th>\n",
       "      <td>blaOXA</td>\n",
       "      <td>Acinetobacter bereziniae Nec blaOXA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4217</th>\n",
       "      <td>blaPDC</td>\n",
       "      <td>Pseudomonas aeruginosa 163613 blaPDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1131</th>\n",
       "      <td>blaADC</td>\n",
       "      <td>Acinetobacter pittii 56 GEIH blaADC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>blaADC</td>\n",
       "      <td>Acinetobacter nosocomialis 12A183 blaADC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2890</th>\n",
       "      <td>blaOXA</td>\n",
       "      <td>Acinetobacter baumannii U21-Benz-S1-1 blaOXA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7372</th>\n",
       "      <td>sul2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4373</th>\n",
       "      <td>blaPDC</td>\n",
       "      <td>Pseudomonas aeruginosa 1800176 blaPDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5726</th>\n",
       "      <td>dfrA1</td>\n",
       "      <td>Vibrio cholerae non-O1/non-O139 dfrA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>aac(6')</td>\n",
       "      <td>Yersinia mollaretii FE82747 aac(6')</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_family                                        refseq_gene\n",
       "4850      blaSHV                 Klebsiella pneumoniae 90088 blaSHV\n",
       "3156      blaOXA                Acinetobacter bereziniae Nec blaOXA\n",
       "4217      blaPDC               Pseudomonas aeruginosa 163613 blaPDC\n",
       "1131      blaADC                Acinetobacter pittii 56 GEIH blaADC\n",
       "1183      blaADC           Acinetobacter nosocomialis 12A183 blaADC\n",
       "2890      blaOXA       Acinetobacter baumannii U21-Benz-S1-1 blaOXA\n",
       "7372        sul2  Salmonella enterica subsp. enterica serovar Ty...\n",
       "4373      blaPDC              Pseudomonas aeruginosa 1800176 blaPDC\n",
       "5726       dfrA1              Vibrio cholerae non-O1/non-O139 dfrA1\n",
       "77       aac(6')                Yersinia mollaretii FE82747 aac(6')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene is only in one column -- everything Nan is dropped \n",
    "# Gene_family was included in orginal dataframe and matches with the found gene -- Should be okay: expert knowledge \n",
    "df.dropna(subset=\"refseq_gene\", inplace=True)\n",
    "df[[\"gene_family\", \"refseq_gene\"]].sample(10, random_state=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "296c701f-05b6-4600-a27e-f27dbf9030e0",
   "metadata": {},
   "source": [
    "## Search for connection to Wikidata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2527c9e3-e9db-4ddf-8dd0-b16d7f24c3d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "\n",
    "sparql = SPARQLWrapper(\"https://query.wikidata.org/sparql\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "792350f6-ad93-4454-a15f-938e16d8c96a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def search_parent_taxon(df_row: pd.Series) -> str:\n",
    "    \"\"\"\n",
    "    Searches in wikidata for a species of bacterium which matches the two first words of strain\n",
    "    Sometime we have abbreviations for example Achromobacter sp. -- nobody knows if this is rather Achromobacter spanius or Achromobacter spiritinus -- This is all rather unusable\n",
    "    \"\"\"\n",
    "    parent = df_row[\"strain\"].split()[:2]\n",
    "    print(df_row[\"strain\"])\n",
    "    abbreviation = False\n",
    "    for i, word in enumerate(parent): \n",
    "        if word[-1] == \".\": \n",
    "            abbreviation = True\n",
    "            parent[i] = word[:-1]\n",
    "    parent = \" \".join(parent).lower()\n",
    "    query = f\"\"\"SELECT ?item ?itemLabel ?itemDescription\n",
    "    WHERE {{\n",
    "      ?item rdfs:label ?label;\n",
    "            schema:description \"species of bacterium\"@en.\n",
    "      \n",
    "      FILTER(LANG(?label) = \"en\" && CONTAINS(LCASE(?label), \"{parent}\"))\n",
    "      \n",
    "      SERVICE wikibase:label {{ bd:serviceParam wikibase:language \"[AUTO_LANGUAGE],en\". }}\n",
    "    }}\n",
    "    LIMIT 10\n",
    "    \"\"\"\n",
    "    sparql.setQuery(query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    while True:\n",
    "        try:\n",
    "            results = sparql.query().convert().get(\"results\").get(\"bindings\")\n",
    "            break\n",
    "        except (urllib.error.HTTPError, urllib.request.HTTPError): \n",
    "            time.sleep(5)\n",
    "        except: \n",
    "            pass\n",
    "    results = [(item.get(\"item\").get(\"value\"), item.get(\"itemLabel\").get(\"value\")) for item in results]\n",
    "    if not results:\n",
    "        return None\n",
    "    if abbreviation: \n",
    "        return [res[0] for res in results] if len(results) > 1 else results[0][0]\n",
    "    else: \n",
    "        for res in results: \n",
    "            if res[1].lower() in parent: \n",
    "                return res[0]\n",
    "\n",
    "\n",
    "get_from_wikidata_switch = False\n",
    "if get_from_wikidata_switch:\n",
    "    # Will take ~ 2h\n",
    "    df[\"parent_taxon\"] = df.apply(search_parent_taxon, axis=1)\n",
    "    df.to_csv(\"resistance_df2.csv\", index=False)\n",
    "else: \n",
    "    df = pd.read_csv(\"resistance_df2.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "853ce8b4-8c71-4427-b416-ce4d0bfeaabb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>product_name</th>\n",
       "      <th>refseq_organism</th>\n",
       "      <th>strain</th>\n",
       "      <th>parent_taxon</th>\n",
       "      <th>refseq_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1420</th>\n",
       "      <td>extended-spectrum class A beta-lactamase CTX-M-74</td>\n",
       "      <td>Enterobacter cloacae</td>\n",
       "      <td>Enterobacter cloacae JF216</td>\n",
       "      <td>http://www.wikidata.org/entity/Q4038096</td>\n",
       "      <td>Enterobacter cloacae JF216 blaCTX-M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5304</th>\n",
       "      <td>tetracycline efflux MFS transporter Tet(L)</td>\n",
       "      <td>Latilactobacillus sakei</td>\n",
       "      <td>Latilactobacillus sakei Rits9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lactobacillus sakei Rits9 pLS55 tet(L)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1453</th>\n",
       "      <td>class C beta-lactamase DHA-18</td>\n",
       "      <td>Morganella morganii</td>\n",
       "      <td>Morganella morganii 984080</td>\n",
       "      <td>http://www.wikidata.org/entity/Q2696880</td>\n",
       "      <td>Morganella morganii 984080 blaDHA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3415</th>\n",
       "      <td>class C beta-lactamase PDC-156</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Pseudomonas aeruginosa 1231451</td>\n",
       "      <td>http://www.wikidata.org/entity/Q31856</td>\n",
       "      <td>Pseudomonas aeruginosa 1231451 blaPDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>extended-spectrum class C beta-lactamase ADC-256</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>Acinetobacter baumannii 20A3025</td>\n",
       "      <td>http://www.wikidata.org/entity/Q3241189</td>\n",
       "      <td>Acinetobacter baumannii 20A3025 blaADC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           product_name  \\\n",
       "1420  extended-spectrum class A beta-lactamase CTX-M-74   \n",
       "5304         tetracycline efflux MFS transporter Tet(L)   \n",
       "1453                      class C beta-lactamase DHA-18   \n",
       "3415                     class C beta-lactamase PDC-156   \n",
       "780    extended-spectrum class C beta-lactamase ADC-256   \n",
       "\n",
       "              refseq_organism                           strain  \\\n",
       "1420     Enterobacter cloacae       Enterobacter cloacae JF216   \n",
       "5304  Latilactobacillus sakei    Latilactobacillus sakei Rits9   \n",
       "1453      Morganella morganii       Morganella morganii 984080   \n",
       "3415   Pseudomonas aeruginosa   Pseudomonas aeruginosa 1231451   \n",
       "780   Acinetobacter baumannii  Acinetobacter baumannii 20A3025   \n",
       "\n",
       "                                 parent_taxon  \\\n",
       "1420  http://www.wikidata.org/entity/Q4038096   \n",
       "5304                                      NaN   \n",
       "1453  http://www.wikidata.org/entity/Q2696880   \n",
       "3415    http://www.wikidata.org/entity/Q31856   \n",
       "780   http://www.wikidata.org/entity/Q3241189   \n",
       "\n",
       "                                 refseq_gene  \n",
       "1420     Enterobacter cloacae JF216 blaCTX-M  \n",
       "5304  Lactobacillus sakei Rits9 pLS55 tet(L)  \n",
       "1453       Morganella morganii 984080 blaDHA  \n",
       "3415   Pseudomonas aeruginosa 1231451 blaPDC  \n",
       "780   Acinetobacter baumannii 20A3025 blaADC  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The dataframe now contains a column called parent taxon which leads to the corresponding wikidata species of bacterium \n",
    "# Some parent taxon could not be found -- will be dropped later \n",
    "df[[\"product_name\", \"refseq_organism\",  \"strain\", \"parent_taxon\", \"refseq_gene\"]].sample(5, random_state=9)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f01c3226-ff91-439c-8c76-5cae46a885f4",
   "metadata": {},
   "source": [
    "### Handle special cases\n",
    "\n",
    "For example: Escherichia coli K-12 \n",
    "I found this randomly -- could be more -- expert knowledge required"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2042e34f-6423-4a04-8e40-4f4896388404",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['allele', 'gene_family', 'whitelisted_taxa', 'product_name', 'scope',\n",
       "       'type', 'subtype', 'class', 'subclass', 'refseq_protein_accession',\n",
       "       'refseq_nucleotide_accession', 'curated_refseq_start',\n",
       "       'genbank_protein_accession', 'genbank_nucleotide_accession',\n",
       "       'genbank_strand', 'genbank_start', 'genbank_stop', 'refseq_strand',\n",
       "       'refseq_start', 'refseq_stop', 'pubmed_reference', 'blacklisted_taxa',\n",
       "       'synonyms', 'hierarchy_node', 'db_version', 'refseq_parent_taxon',\n",
       "       'refseq_protein', 'refseq_parent_taxon2', 'refseq_gene',\n",
       "       'refseq_protein2', 'refseq_genome', 'refseq_organism', 'refseq_tax_id',\n",
       "       'genbank_organism', 'genbank_strain', 'genbank_organism2',\n",
       "       'genbank_strain2', 'genbank_tax_id', 'strain', 'parent_taxon'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "73494bab-50f0-4b35-8b52-9377d4432700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele</th>\n",
       "      <th>gene_family</th>\n",
       "      <th>whitelisted_taxa</th>\n",
       "      <th>product_name</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>refseq_protein_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>refseq_genome</th>\n",
       "      <th>refseq_organism</th>\n",
       "      <th>refseq_tax_id</th>\n",
       "      <th>genbank_organism</th>\n",
       "      <th>genbank_strain</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_tax_id</th>\n",
       "      <th>strain</th>\n",
       "      <th>parent_taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aph(4)-Ia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside O-phosphotransferase APH(4)-Ia</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>HYGROMYCIN</td>\n",
       "      <td>WP_000742814.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>taxon:83333</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>K-12</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>K-12</td>\n",
       "      <td>taxon:83333</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>https://www.wikidata.org/entity/Q21399437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4443</th>\n",
       "      <td>NaN</td>\n",
       "      <td>catA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>type A-2 chloramphenicol O-acetyltransferase C...</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>PHENICOL</td>\n",
       "      <td>CHLORAMPHENICOL</td>\n",
       "      <td>WP_012477888.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>taxon:83333</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>K-12</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>K-12</td>\n",
       "      <td>taxon:83333</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "      <td>https://www.wikidata.org/entity/Q21399437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     allele gene_family whitelisted_taxa  \\\n",
       "459     NaN   aph(4)-Ia              NaN   \n",
       "4443    NaN       catA2              NaN   \n",
       "\n",
       "                                           product_name scope type subtype  \\\n",
       "459       aminoglycoside O-phosphotransferase APH(4)-Ia  core  AMR     AMR   \n",
       "4443  type A-2 chloramphenicol O-acetyltransferase C...  core  AMR     AMR   \n",
       "\n",
       "               class         subclass refseq_protein_accession  ...  \\\n",
       "459   AMINOGLYCOSIDE       HYGROMYCIN           WP_000742814.1  ...   \n",
       "4443        PHENICOL  CHLORAMPHENICOL           WP_012477888.1  ...   \n",
       "\n",
       "     refseq_genome        refseq_organism refseq_tax_id  \\\n",
       "459            NaN  Escherichia coli K-12   taxon:83333   \n",
       "4443           NaN  Escherichia coli K-12   taxon:83333   \n",
       "\n",
       "           genbank_organism genbank_strain      genbank_organism2  \\\n",
       "459   Escherichia coli K-12           K-12  Escherichia coli K-12   \n",
       "4443  Escherichia coli K-12           K-12  Escherichia coli K-12   \n",
       "\n",
       "      genbank_strain2 genbank_tax_id                 strain  \\\n",
       "459              K-12    taxon:83333  Escherichia coli K-12   \n",
       "4443             K-12    taxon:83333  Escherichia coli K-12   \n",
       "\n",
       "                                   parent_taxon  \n",
       "459   https://www.wikidata.org/entity/Q21399437  \n",
       "4443  https://www.wikidata.org/entity/Q21399437  \n",
       "\n",
       "[2 rows x 40 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"parent_taxon\"] = np.where((df[\"genbank_organism2\"].str.lower().str.contains(\"escherichia coli\")) & (df[\"genbank_strain2\"] == \"K-12\"), \n",
    "                             \"https://www.wikidata.org/entity/Q21399437\", \n",
    "                             df[\"parent_taxon\"])\n",
    "df[df[\"parent_taxon\"] == \"https://www.wikidata.org/entity/Q21399437\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03292861-a0ca-4c1b-b147-01600ebbb647",
   "metadata": {},
   "source": [
    "## Drop useless data instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9b69d3fe-69f8-4d03-9f55-64ad5cd3ec9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop rows where no wikidata parent taxon (species of bacterium) was found \n",
    "df = df.dropna(subset=\"parent_taxon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2bd07b8b-932e-4996-855d-d9ffa590d463",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>parent_taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Streptomyces sp. 769</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Streptomyces sp. GBA 94-10 4N24</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Streptomyces sp. SPB78</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Streptomyces sp. NRRL S-1831</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Streptomyces sp. MBRL 601</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Streptomyces sp. M10</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Streptomyces sp. KE1</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Streptomyces sp. NRRL F-4711</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Streptomyces sp. NRRL F-4707</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>Streptomyces sp. NRRL S-1868</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1954</th>\n",
       "      <td>Streptomyces sp. NRRL F-5053</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2734</th>\n",
       "      <td>Achromobacter sp. 2015clin2ach</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q16973926', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2735</th>\n",
       "      <td>Achromobacter sp. 2015clin4ach</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q16973926', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4300</th>\n",
       "      <td>Achromobacter sp. sallbros1</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q16973926', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4430</th>\n",
       "      <td>Streptomyces sp. MBRL 601</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4501</th>\n",
       "      <td>Clostridium sp. AF37-7</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4698</th>\n",
       "      <td>Lactobacillus sp. PC121B</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q3825343', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5180</th>\n",
       "      <td>Streptomyces sp. WAC 01438 WAC1438</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5196</th>\n",
       "      <td>Streptomyces sp. WAC4747</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q60748845', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5247</th>\n",
       "      <td>Corynebacterium sp. L2-79-05</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q21262825', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5305</th>\n",
       "      <td>Bacillus sp. 24</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q798286', 'ht...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5310</th>\n",
       "      <td>Bacillus sp. R89</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q798286', 'ht...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5405</th>\n",
       "      <td>Clostridium sp. MLG480</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5406</th>\n",
       "      <td>Clostridium sp. MLG080-1</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5407</th>\n",
       "      <td>Clostridium sp. MLG055</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5409</th>\n",
       "      <td>Clostridium sp. MLG101</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5501</th>\n",
       "      <td>Clostridium sp. CCRI-9842</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5537</th>\n",
       "      <td>Clostridium sp. CCRI-9842</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5552</th>\n",
       "      <td>Clostridium sp. CCRI-9842</td>\n",
       "      <td>['http://www.wikidata.org/entity/Q2979669', 'h...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  strain  \\\n",
       "38                  Streptomyces sp. 769   \n",
       "46       Streptomyces sp. GBA 94-10 4N24   \n",
       "52                Streptomyces sp. SPB78   \n",
       "59          Streptomyces sp. NRRL S-1831   \n",
       "74             Streptomyces sp. MBRL 601   \n",
       "80                  Streptomyces sp. M10   \n",
       "82                  Streptomyces sp. KE1   \n",
       "89          Streptomyces sp. NRRL F-4711   \n",
       "90          Streptomyces sp. NRRL F-4707   \n",
       "1953        Streptomyces sp. NRRL S-1868   \n",
       "1954        Streptomyces sp. NRRL F-5053   \n",
       "2734      Achromobacter sp. 2015clin2ach   \n",
       "2735      Achromobacter sp. 2015clin4ach   \n",
       "4300         Achromobacter sp. sallbros1   \n",
       "4430           Streptomyces sp. MBRL 601   \n",
       "4501              Clostridium sp. AF37-7   \n",
       "4698            Lactobacillus sp. PC121B   \n",
       "5180  Streptomyces sp. WAC 01438 WAC1438   \n",
       "5196            Streptomyces sp. WAC4747   \n",
       "5247        Corynebacterium sp. L2-79-05   \n",
       "5305                     Bacillus sp. 24   \n",
       "5310                    Bacillus sp. R89   \n",
       "5405              Clostridium sp. MLG480   \n",
       "5406            Clostridium sp. MLG080-1   \n",
       "5407              Clostridium sp. MLG055   \n",
       "5409              Clostridium sp. MLG101   \n",
       "5501           Clostridium sp. CCRI-9842   \n",
       "5537           Clostridium sp. CCRI-9842   \n",
       "5552           Clostridium sp. CCRI-9842   \n",
       "\n",
       "                                           parent_taxon  \n",
       "38    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "46    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "52    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "59    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "74    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "80    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "82    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "89    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "90    ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "1953  ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "1954  ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "2734  ['http://www.wikidata.org/entity/Q16973926', '...  \n",
       "2735  ['http://www.wikidata.org/entity/Q16973926', '...  \n",
       "4300  ['http://www.wikidata.org/entity/Q16973926', '...  \n",
       "4430  ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "4501  ['http://www.wikidata.org/entity/Q2979669', 'h...  \n",
       "4698  ['http://www.wikidata.org/entity/Q3825343', 'h...  \n",
       "5180  ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "5196  ['http://www.wikidata.org/entity/Q60748845', '...  \n",
       "5247  ['http://www.wikidata.org/entity/Q21262825', '...  \n",
       "5305  ['http://www.wikidata.org/entity/Q798286', 'ht...  \n",
       "5310  ['http://www.wikidata.org/entity/Q798286', 'ht...  \n",
       "5405  ['http://www.wikidata.org/entity/Q2979669', 'h...  \n",
       "5406  ['http://www.wikidata.org/entity/Q2979669', 'h...  \n",
       "5407  ['http://www.wikidata.org/entity/Q2979669', 'h...  \n",
       "5409  ['http://www.wikidata.org/entity/Q2979669', 'h...  \n",
       "5501  ['http://www.wikidata.org/entity/Q2979669', 'h...  \n",
       "5537  ['http://www.wikidata.org/entity/Q2979669', 'h...  \n",
       "5552  ['http://www.wikidata.org/entity/Q2979669', 'h...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# These data instances are made up of abbreviations, which make it unclear to which taxon they belong. \n",
    "# e.g. Streptomyces sp. 769 could belong to Streptomyces sp. myrophorea (Q60748847), Streptomyces spiramyceticus (Q104909301) or Streptomyces sporangiiformans (Q104957131)\n",
    "# Expert Knowledge required -- They also need to be dropped \n",
    "df.loc[:, \"parent_taxon\"] = df[\"parent_taxon\"].astype(str)\n",
    "df.loc[df[\"parent_taxon\"].str.contains(\"\\[|\\]\"), [\"strain\", \"parent_taxon\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3e10600c-2c0b-41fd-8735-362a2ba688af",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[~df[\"parent_taxon\"].str.contains(\"\\[|\\]\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "270d3ebd-fff1-424c-a768-b848cc63ea8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4563"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df) # About half of the data is lost after everything is clearead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "864a45ce-6ff8-4a14-b7e7-6b3f2c0906ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>product_name</th>\n",
       "      <th>refseq_gene</th>\n",
       "      <th>parent_taxon</th>\n",
       "      <th>strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3752</th>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CEPHALOSPORIN</td>\n",
       "      <td>inhibitor-resistant class C beta-lactamase PDC...</td>\n",
       "      <td>Pseudomonas aeruginosa 208176 blaPDC</td>\n",
       "      <td>http://www.wikidata.org/entity/Q31856</td>\n",
       "      <td>Pseudomonas aeruginosa 208176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>KANAMYCIN/TOBRAMYCIN</td>\n",
       "      <td>aminoglycoside 6'-N-acetyltransferase AacA34</td>\n",
       "      <td>Klebsiella pneumoniae KP-PNK-1 aacA34</td>\n",
       "      <td>http://www.wikidata.org/entity/Q132592</td>\n",
       "      <td>Klebsiella pneumoniae KP-PNK-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CEPHALOSPORIN</td>\n",
       "      <td>cephalosporin-hydrolyzing class C beta-lactama...</td>\n",
       "      <td>Enterobacter cloacae 963327 blaACT</td>\n",
       "      <td>http://www.wikidata.org/entity/Q4038096</td>\n",
       "      <td>Enterobacter cloacae 963327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5074</th>\n",
       "      <td>QUINOLONE</td>\n",
       "      <td>QUINOLONE</td>\n",
       "      <td>quinolone resistance pentapeptide repeat prote...</td>\n",
       "      <td>Citrobacter freundii V1 pCFV1 qnrB</td>\n",
       "      <td>http://www.wikidata.org/entity/Q5122842</td>\n",
       "      <td>Citrobacter freundii V1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CEPHALOSPORIN</td>\n",
       "      <td>class C beta-lactamase CMY-124</td>\n",
       "      <td>Citrobacter freundii DNS-2 blaCMY</td>\n",
       "      <td>http://www.wikidata.org/entity/Q5122842</td>\n",
       "      <td>Citrobacter freundii DNS-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               class              subclass  \\\n",
       "3752     BETA-LACTAM         CEPHALOSPORIN   \n",
       "266   AMINOGLYCOSIDE  KANAMYCIN/TOBRAMYCIN   \n",
       "543      BETA-LACTAM         CEPHALOSPORIN   \n",
       "5074       QUINOLONE             QUINOLONE   \n",
       "1095     BETA-LACTAM         CEPHALOSPORIN   \n",
       "\n",
       "                                           product_name  \\\n",
       "3752  inhibitor-resistant class C beta-lactamase PDC...   \n",
       "266        aminoglycoside 6'-N-acetyltransferase AacA34   \n",
       "543   cephalosporin-hydrolyzing class C beta-lactama...   \n",
       "5074  quinolone resistance pentapeptide repeat prote...   \n",
       "1095                     class C beta-lactamase CMY-124   \n",
       "\n",
       "                                refseq_gene  \\\n",
       "3752   Pseudomonas aeruginosa 208176 blaPDC   \n",
       "266   Klebsiella pneumoniae KP-PNK-1 aacA34   \n",
       "543      Enterobacter cloacae 963327 blaACT   \n",
       "5074     Citrobacter freundii V1 pCFV1 qnrB   \n",
       "1095      Citrobacter freundii DNS-2 blaCMY   \n",
       "\n",
       "                                 parent_taxon                          strain  \n",
       "3752    http://www.wikidata.org/entity/Q31856   Pseudomonas aeruginosa 208176  \n",
       "266    http://www.wikidata.org/entity/Q132592  Klebsiella pneumoniae KP-PNK-1  \n",
       "543   http://www.wikidata.org/entity/Q4038096     Enterobacter cloacae 963327  \n",
       "5074  http://www.wikidata.org/entity/Q5122842         Citrobacter freundii V1  \n",
       "1095  http://www.wikidata.org/entity/Q5122842      Citrobacter freundii DNS-2  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I now have a dataframe which contains antibiotic resistance class / subclass, protein name, gene, species of bacterium (wikidata) and bacterial strain \n",
    "# Could be implemented into wikidata like this \n",
    "df[[\"class\", \"subclass\", \"product_name\", \"refseq_gene\", \"parent_taxon\", \"strain\"]].sample(5, random_state=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6bd2df29-36f0-4dc1-91b0-a5835a47a632",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NCBI taxonomy ID is in all data instances regardsless of access via refseq or genbank the same \n",
    "any(df[\"genbank_tax_id\"] != df[\"refseq_tax_id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fe83dbd6-3387-4ca6-8f07-dd9981f40e74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_tax_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5129</th>\n",
       "      <td>Citrobacter braakii 107</td>\n",
       "      <td>Citrobacter braakii</td>\n",
       "      <td>107</td>\n",
       "      <td>taxon:57706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2998</th>\n",
       "      <td>Acinetobacter baumannii 16-02P46T-1</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>16-02P46T-1</td>\n",
       "      <td>taxon:470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3013</th>\n",
       "      <td>Acinetobacter baumannii 17A1872</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>17A1872</td>\n",
       "      <td>taxon:470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4113</th>\n",
       "      <td>Klebsiella pneumoniae 1409130</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>1409130</td>\n",
       "      <td>taxon:573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Serratia marcescens W2.3</td>\n",
       "      <td>Serratia marcescens W2.3</td>\n",
       "      <td>W2.3</td>\n",
       "      <td>taxon:1218513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3965</th>\n",
       "      <td>Pseudomonas libanensis DSM 17149</td>\n",
       "      <td>Pseudomonas libanensis</td>\n",
       "      <td>DSM 17149</td>\n",
       "      <td>taxon:75588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>865</th>\n",
       "      <td>Acinetobacter baumannii 23A3701</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>23A3701</td>\n",
       "      <td>taxon:470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4389</th>\n",
       "      <td>Vibrio alginolyticus Vb1833</td>\n",
       "      <td>Vibrio alginolyticus</td>\n",
       "      <td>Vb1833</td>\n",
       "      <td>taxon:663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>taxon:90371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3497</th>\n",
       "      <td>Pseudomonas aeruginosa 163604</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>163604</td>\n",
       "      <td>taxon:287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 strain  \\\n",
       "5129                            Citrobacter braakii 107   \n",
       "2998                Acinetobacter baumannii 16-02P46T-1   \n",
       "3013                    Acinetobacter baumannii 17A1872   \n",
       "4113                      Klebsiella pneumoniae 1409130   \n",
       "30                             Serratia marcescens W2.3   \n",
       "3965                   Pseudomonas libanensis DSM 17149   \n",
       "865                     Acinetobacter baumannii 23A3701   \n",
       "4389                        Vibrio alginolyticus Vb1833   \n",
       "445   Salmonella enterica subsp. enterica serovar Ty...   \n",
       "3497                      Pseudomonas aeruginosa 163604   \n",
       "\n",
       "                                      genbank_organism2 genbank_strain2  \\\n",
       "5129                                Citrobacter braakii             107   \n",
       "2998                            Acinetobacter baumannii     16-02P46T-1   \n",
       "3013                            Acinetobacter baumannii         17A1872   \n",
       "4113                              Klebsiella pneumoniae         1409130   \n",
       "30                             Serratia marcescens W2.3            W2.3   \n",
       "3965                             Pseudomonas libanensis       DSM 17149   \n",
       "865                             Acinetobacter baumannii         23A3701   \n",
       "4389                               Vibrio alginolyticus          Vb1833   \n",
       "445   Salmonella enterica subsp. enterica serovar Ty...             NaN   \n",
       "3497                             Pseudomonas aeruginosa          163604   \n",
       "\n",
       "     genbank_tax_id  \n",
       "5129    taxon:57706  \n",
       "2998      taxon:470  \n",
       "3013      taxon:470  \n",
       "4113      taxon:573  \n",
       "30    taxon:1218513  \n",
       "3965    taxon:75588  \n",
       "865       taxon:470  \n",
       "4389      taxon:663  \n",
       "445     taxon:90371  \n",
       "3497      taxon:287  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# But the NCBI taxonomy ID has not always enough depth -- e.g. line 3664 taxon 470 -> https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi leads to acinetobacter baumannii \n",
    "# but not to the corresponding strain (16-02P46T-1) which does not exist yet. \n",
    "# Line 59 Serratia marcescens W2.3 leads to the wanted result: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi --> 1218513\n",
    "# What sould be done here? Expert knowledge required -- I don't feel well implementing the data without the correct taxonomy ID\n",
    "# I will drop rows where strain and genbank_organism2 don't match - basically undo combination step further up\n",
    "df[[\"strain\", \"genbank_organism2\", \"genbank_strain2\", \"genbank_tax_id\"]].sample(10, random_state=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "699aaf1e-a5aa-4555-b7cc-4f0062221d18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "312\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_tax_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Acinetobacter baumannii 146457</td>\n",
       "      <td>Acinetobacter baumannii 146457</td>\n",
       "      <td>146457</td>\n",
       "      <td>taxon:1310623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Acinetobacter baumannii TG02011</td>\n",
       "      <td>Acinetobacter baumannii TG02011</td>\n",
       "      <td>TG02011</td>\n",
       "      <td>taxon:1315135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4954</th>\n",
       "      <td>Exiguobacterium sp. S3-2</td>\n",
       "      <td>Exiguobacterium sp. S3-2</td>\n",
       "      <td>S3-2</td>\n",
       "      <td>taxon:1389960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>Enterococcus faecium SD3B-2</td>\n",
       "      <td>Enterococcus faecium SD3B-2</td>\n",
       "      <td>SD3B-2</td>\n",
       "      <td>taxon:1244155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>Enterococcus hirae ATCC 9790</td>\n",
       "      <td>Enterococcus hirae ATCC 9790</td>\n",
       "      <td>ATCC 9790</td>\n",
       "      <td>taxon:768486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1597</th>\n",
       "      <td>Elizabethkingia anophelis NUHP1</td>\n",
       "      <td>Elizabethkingia anophelis NUHP1</td>\n",
       "      <td>NUHP1</td>\n",
       "      <td>taxon:1338011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5355</th>\n",
       "      <td>Bifidobacterium longum subsp. longum F8</td>\n",
       "      <td>Bifidobacterium longum subsp. longum F8</td>\n",
       "      <td>F8</td>\n",
       "      <td>taxon:722911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5194</th>\n",
       "      <td>Nocardia farcinica IFM 10152</td>\n",
       "      <td>Nocardia farcinica IFM 10152</td>\n",
       "      <td>IFM 10152</td>\n",
       "      <td>taxon:247156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1022</th>\n",
       "      <td>Vibrio parahaemolyticus S105</td>\n",
       "      <td>Vibrio parahaemolyticus S105</td>\n",
       "      <td>S105</td>\n",
       "      <td>taxon:1394641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Serratia marcescens MC620</td>\n",
       "      <td>Serratia marcescens MC620</td>\n",
       "      <td>MC620</td>\n",
       "      <td>taxon:1333585</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       strain  \\\n",
       "138            Acinetobacter baumannii 146457   \n",
       "269           Acinetobacter baumannii TG02011   \n",
       "4954                 Exiguobacterium sp. S3-2   \n",
       "159               Enterococcus faecium SD3B-2   \n",
       "161              Enterococcus hirae ATCC 9790   \n",
       "1597          Elizabethkingia anophelis NUHP1   \n",
       "5355  Bifidobacterium longum subsp. longum F8   \n",
       "5194             Nocardia farcinica IFM 10152   \n",
       "1022             Vibrio parahaemolyticus S105   \n",
       "78                  Serratia marcescens MC620   \n",
       "\n",
       "                            genbank_organism2 genbank_strain2 genbank_tax_id  \n",
       "138            Acinetobacter baumannii 146457          146457  taxon:1310623  \n",
       "269           Acinetobacter baumannii TG02011         TG02011  taxon:1315135  \n",
       "4954                 Exiguobacterium sp. S3-2            S3-2  taxon:1389960  \n",
       "159               Enterococcus faecium SD3B-2          SD3B-2  taxon:1244155  \n",
       "161              Enterococcus hirae ATCC 9790       ATCC 9790   taxon:768486  \n",
       "1597          Elizabethkingia anophelis NUHP1           NUHP1  taxon:1338011  \n",
       "5355  Bifidobacterium longum subsp. longum F8              F8   taxon:722911  \n",
       "5194             Nocardia farcinica IFM 10152       IFM 10152   taxon:247156  \n",
       "1022             Vibrio parahaemolyticus S105            S105  taxon:1394641  \n",
       "78                  Serratia marcescens MC620           MC620  taxon:1333585  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# These would be the data instances where I feel comfortable, including into Wikidata, because I have the correct NCBI taxonomy ID for others to check. \n",
    "# Now there are only 312 instances left \n",
    "df = df.loc[df.apply(lambda x: str(x[\"genbank_strain2\"]) in x[\"genbank_organism2\"], axis=1), :]\n",
    "print(len(df))\n",
    "df[[\"strain\", \"genbank_organism2\", \"genbank_strain2\", \"genbank_tax_id\"]].sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9d11636d-fb29-4502-a120-fea3b5de5902",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"resistance_df3.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "642e3a0d-1e96-42c4-a210-6d0cb1eb9b40",
   "metadata": {},
   "source": [
    "## Results\n",
    "\n",
    "After selecting and combining the best data sources (see the commented code above), a dataframe remains containing information about the protein, its encoding gene, and the bacterial strain. To establish a connection with Wikidata, the first two words of the bacterial strain (often the name of the corresponding bacterial species) are used in a SPARQL Wikidata query to find the associated bacterial species or its Wikidata qualifier. In addition to the respective names, identifiers are available to link each name with the NCBI database. This connection is crucial to provide users with more comprehensive information and allow experts to make improvements.\n",
    "\n",
    "Furthermore, data is cleaned if there are missing or conflicting details at critical points. The most extensive cleaning, associated with the greatest data loss, occurs when it was observed that the found NCBI Taxonomy ID does not always refer to the bacterial strain but rather to higher taxa. This is unacceptable for Wikidata implementation, leading to the removal of such instances.\n",
    "\n",
    "In the final step, the data is implemented in Wikidata. Using \"pywikibots,\" a connection is established. Firstly, if not already present, the bacterial strain is implemented with the NCBI taxonomy ID. Subsequently, the identified gene is implemented, referencing the bacterial strain. Finally, the gene encodes a protein that makes the bacterium antibiotic-resistant. This protein is also implemented with a corresponding reference to the gene and the Quick-Go reference \"response to antibiotic\" (GO:0046677).\n",
    "\n",
    "As of the submission deadline, work is ongoing on the Wikidata implementation. For instance, the bacterial strain Serratia marcescens VGH107 (Q124664471) has already been implemented. Additional data will be integrated in the coming days."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "3d985c62-9f7b-4ced-8e71-8919349237d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pywikibot\n",
    "\n",
    "\n",
    "class WikidataAdder: \n",
    "\n",
    "    site = pywikibot.Site(\"wikidata\")\n",
    "    repo = site.data_repository()\n",
    "\n",
    "    def __init__(self, df_row: pd.Series, sim: bool): \n",
    "        self.df_row = df_row \n",
    "        self.strain_page: pywikibot.ItemPage = None\n",
    "        self.sim = sim\n",
    "\n",
    "    def return_strain_page(self): \n",
    "        if self.sim:\n",
    "            return \"\"\n",
    "        return self.strain_page.getID()\n",
    "\n",
    "    def create_strain(self) -> str | bool: \n",
    "        if self.sim: \n",
    "            print(\"Label: \" + self.df_row[\"genbank_organism2\"])\n",
    "            print(\"Description: \" + \"bacterial strain\")\n",
    "            print(\"Alias: \" + self.df_row[\"genbank_organism2\"].split()[-1])\n",
    "        else:\n",
    "            query = f\"\"\"\n",
    "                        SELECT ?item WHERE \n",
    "                            {{?item rdfs:label \"{self.df_row[\"genbank_organism2\"]}\"@en}}\n",
    "                    \"\"\"\n",
    "            sparql.setQuery(query)\n",
    "            sparql.setReturnFormat(JSON)\n",
    "            results = sparql.query().convert().get(\"results\").get(\"bindings\")\n",
    "            print(results)\n",
    "            if len(results) >= 2: \n",
    "                print(\"more than one item found\")\n",
    "                self.sim = True \n",
    "                return \n",
    "            if len(results) == 1:\n",
    "                self.strain_page = pywikibot.ItemPage(self.repo, results[0].get(\"item\").get(\"value\").split(\"/\")[-1])\n",
    "            elif not results: \n",
    "                # self.strain_page = pywikibot.ItemPage(WikidataAdder.site)\n",
    "                # self.strain_page.editLabels({\"en\": self.df_row[\"genbank_organism2\"]}, summary=\"Setting new label\")\n",
    "                # self.strain_page.editDescriptions({\"en\": \"bacterial strain\"}, summary=\"Setting new description\")\n",
    "                # self.strain_page.editAliases({\"en\": [self.df_row[\"genbank_organism2\"].split()[-1]]}, summary=\"Setting new alias\")\n",
    "                pass\n",
    "            print(self.strain_page.get()[\"labels\"])\n",
    "            print(self.strain_page.get()[\"descriptions\"])\n",
    "            print(self.strain_page.get()[\"aliases\"][\"en\"])\n",
    "        \n",
    "\n",
    "    def add_instance_of_strain(self):\n",
    "        if not self.sim:\n",
    "            claim = pywikibot.Claim(WikidataAdder.repo, u\"P31\")\n",
    "            target = pywikibot.ItemPage(WikidataAdder.repo, u\"Q855769\")\n",
    "            claim.setTarget(target)\n",
    "            self.strain_page.addClaim(claim, summary=u'Adding claim')\n",
    "\n",
    "    def add_taxon_name(self): \n",
    "        if not self.sim:\n",
    "            stringclaim = pywikibot.Claim(WikidataAdder.repo, u'P225')\n",
    "            stringclaim.setTarget(self.df_row[\"genbank_organism2\"])\n",
    "            self.strain_page.addClaim(stringclaim, summary=u'Adding taxon name')\n",
    "\n",
    "    def add_parent_taxon(self):\n",
    "        if not self.sim:\n",
    "            claim = pywikibot.Claim(self.repo, u\"P171\")\n",
    "            target = pywikibot.ItemPage(self.repo, self.df_row[\"parent_taxon\"].split(\"/\")[-1])\n",
    "            claim.setTarget(target)\n",
    "            self.strain_page.addClaim(claim, summary=u'Adding parent taxon')\n",
    "\n",
    "\n",
    "def wikidata_wrapper(df_row: pd.Series) -> pd.Series: \n",
    "    wa = WikidataAdder(df_row, sim=False) # Set sim to True unless you really want to include data into wikidata\n",
    "    wa.create_strain()\n",
    "    return \"123\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "a2daadd3-ea51-4fa5-878e-be65d2e8caf6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664031'}}]\n",
      "<class 'pywikibot.page._collections.LanguageDict'>({'en': 'Pseudomonas aeruginosa PA38182'})\n",
      "<class 'pywikibot.page._collections.LanguageDict'>({'en': 'bacterial strain'})\n",
      "['PA38182']\n",
      "[{'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124663344'}}, {'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664085'}}]\n",
      "more than one item found\n",
      "[{'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q21102987'}}, {'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664088'}}]\n",
      "more than one item found\n",
      "[{'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q21398890'}}, {'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664402'}}, {'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664429'}}]\n",
      "more than one item found\n",
      "[{'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q21398562'}}, {'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664441'}}]\n",
      "more than one item found\n",
      "[{'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664454'}}]\n",
      "<class 'pywikibot.page._collections.LanguageDict'>({'en': 'Yersinia mollaretii ATCC 43969'})\n",
      "<class 'pywikibot.page._collections.LanguageDict'>({'en': 'bacterial strain'})\n",
      "['43969']\n",
      "[{'item': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q124664466'}}]\n",
      "<class 'pywikibot.page._collections.LanguageDict'>({'en': 'Serratia marcescens W2.3'})\n",
      "<class 'pywikibot.page._collections.LanguageDict'>({'en': 'bacterial strain'})\n",
      "['W2.3']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele</th>\n",
       "      <th>gene_family</th>\n",
       "      <th>whitelisted_taxa</th>\n",
       "      <th>product_name</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>refseq_protein_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>refseq_organism</th>\n",
       "      <th>refseq_tax_id</th>\n",
       "      <th>genbank_organism</th>\n",
       "      <th>genbank_strain</th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>genbank_strain2</th>\n",
       "      <th>genbank_tax_id</th>\n",
       "      <th>strain</th>\n",
       "      <th>parent_taxon</th>\n",
       "      <th>strain_wd_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aac(2')-I(A267)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside N-acetyltransferase AAC(2')-I(A...</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>GENTAMICIN/TOBRAMYCIN</td>\n",
       "      <td>WP_025297907.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>taxon:1407059</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>PA38182</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>PA38182</td>\n",
       "      <td>taxon:1407059</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>http://www.wikidata.org/entity/Q31856</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aac(2')-IIb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>kasugamycin N-acetyltransferase AAC(2')-IIb</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>KASUGAMYCIN</td>\n",
       "      <td>WP_071224044.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>taxon:1120679</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>LC231</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>LC231</td>\n",
       "      <td>taxon:1120679</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>http://www.wikidata.org/entity/Q26270468</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aac(2')-Ic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside N-acetyltransferase AAC(2')-Ic</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>GENTAMICIN/TOBRAMYCIN</td>\n",
       "      <td>WP_003899880.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>taxon:83332</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>H37Rv</td>\n",
       "      <td>taxon:83332</td>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>http://www.wikidata.org/entity/Q130971</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aac(3)-Ig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside N-acetyltransferase AAC(3)-Ig</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>GENTAMICIN</td>\n",
       "      <td>WP_011468318.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Saccharophagus degradans 2-40</td>\n",
       "      <td>taxon:203122</td>\n",
       "      <td>Saccharophagus degradans 2-40</td>\n",
       "      <td>2-40</td>\n",
       "      <td>Saccharophagus degradans 2-40</td>\n",
       "      <td>2-40</td>\n",
       "      <td>taxon:203122</td>\n",
       "      <td>Saccharophagus degradans 2-40</td>\n",
       "      <td>http://www.wikidata.org/entity/Q7396606</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aac(3)-Ii</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside N-acetyltransferase AAC(3)-Ii</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>GENTAMICIN</td>\n",
       "      <td>WP_011540937.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Sphingopyxis alaskensis RB2256</td>\n",
       "      <td>taxon:317655</td>\n",
       "      <td>Sphingopyxis alaskensis RB2256</td>\n",
       "      <td>RB2256</td>\n",
       "      <td>Sphingopyxis alaskensis RB2256</td>\n",
       "      <td>RB2256</td>\n",
       "      <td>taxon:317655</td>\n",
       "      <td>Sphingopyxis alaskensis RB2256</td>\n",
       "      <td>http://www.wikidata.org/entity/Q21324563</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aac(6')</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside 6'-N-acetyltransferase</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>WP_004874306.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Yersinia mollaretii ATCC 43969</td>\n",
       "      <td>taxon:349967</td>\n",
       "      <td>Yersinia mollaretii ATCC 43969</td>\n",
       "      <td>ATCC 43969</td>\n",
       "      <td>Yersinia mollaretii ATCC 43969</td>\n",
       "      <td>ATCC 43969</td>\n",
       "      <td>taxon:349967</td>\n",
       "      <td>Yersinia mollaretii ATCC 43969</td>\n",
       "      <td>http://www.wikidata.org/entity/Q16994539</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NaN</td>\n",
       "      <td>aac(6')</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aminoglycoside 6'-N-acetyltransferase</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>AMINOGLYCOSIDE</td>\n",
       "      <td>WP_019453091.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Serratia marcescens W2.3</td>\n",
       "      <td>taxon:1218513</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serratia marcescens W2.3</td>\n",
       "      <td>W2.3</td>\n",
       "      <td>taxon:1218513</td>\n",
       "      <td>Serratia marcescens W2.3</td>\n",
       "      <td>http://www.wikidata.org/entity/Q140004</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows Ã— 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   allele      gene_family whitelisted_taxa  \\\n",
       "0     NaN  aac(2')-I(A267)              NaN   \n",
       "2     NaN      aac(2')-IIb              NaN   \n",
       "4     NaN       aac(2')-Ic              NaN   \n",
       "20    NaN        aac(3)-Ig              NaN   \n",
       "21    NaN        aac(3)-Ii              NaN   \n",
       "27    NaN          aac(6')              NaN   \n",
       "30    NaN          aac(6')              NaN   \n",
       "\n",
       "                                         product_name scope type subtype  \\\n",
       "0   aminoglycoside N-acetyltransferase AAC(2')-I(A...  core  AMR     AMR   \n",
       "2         kasugamycin N-acetyltransferase AAC(2')-IIb  core  AMR     AMR   \n",
       "4       aminoglycoside N-acetyltransferase AAC(2')-Ic  core  AMR     AMR   \n",
       "20       aminoglycoside N-acetyltransferase AAC(3)-Ig  core  AMR     AMR   \n",
       "21       aminoglycoside N-acetyltransferase AAC(3)-Ii  core  AMR     AMR   \n",
       "27              aminoglycoside 6'-N-acetyltransferase  core  AMR     AMR   \n",
       "30              aminoglycoside 6'-N-acetyltransferase  core  AMR     AMR   \n",
       "\n",
       "             class               subclass refseq_protein_accession  ...  \\\n",
       "0   AMINOGLYCOSIDE  GENTAMICIN/TOBRAMYCIN           WP_025297907.1  ...   \n",
       "2   AMINOGLYCOSIDE            KASUGAMYCIN           WP_071224044.1  ...   \n",
       "4   AMINOGLYCOSIDE  GENTAMICIN/TOBRAMYCIN           WP_003899880.1  ...   \n",
       "20  AMINOGLYCOSIDE             GENTAMICIN           WP_011468318.1  ...   \n",
       "21  AMINOGLYCOSIDE             GENTAMICIN           WP_011540937.1  ...   \n",
       "27  AMINOGLYCOSIDE         AMINOGLYCOSIDE           WP_004874306.1  ...   \n",
       "30  AMINOGLYCOSIDE         AMINOGLYCOSIDE           WP_019453091.1  ...   \n",
       "\n",
       "                     refseq_organism  refseq_tax_id  \\\n",
       "0     Pseudomonas aeruginosa PA38182  taxon:1407059   \n",
       "2            Paenibacillus sp. LC231  taxon:1120679   \n",
       "4   Mycobacterium tuberculosis H37Rv    taxon:83332   \n",
       "20     Saccharophagus degradans 2-40   taxon:203122   \n",
       "21    Sphingopyxis alaskensis RB2256   taxon:317655   \n",
       "27    Yersinia mollaretii ATCC 43969   taxon:349967   \n",
       "30          Serratia marcescens W2.3  taxon:1218513   \n",
       "\n",
       "                    genbank_organism genbank_strain  \\\n",
       "0     Pseudomonas aeruginosa PA38182        PA38182   \n",
       "2            Paenibacillus sp. LC231          LC231   \n",
       "4   Mycobacterium tuberculosis H37Rv          H37Rv   \n",
       "20     Saccharophagus degradans 2-40           2-40   \n",
       "21    Sphingopyxis alaskensis RB2256         RB2256   \n",
       "27    Yersinia mollaretii ATCC 43969     ATCC 43969   \n",
       "30                               NaN            NaN   \n",
       "\n",
       "                   genbank_organism2  genbank_strain2  genbank_tax_id  \\\n",
       "0     Pseudomonas aeruginosa PA38182          PA38182   taxon:1407059   \n",
       "2            Paenibacillus sp. LC231            LC231   taxon:1120679   \n",
       "4   Mycobacterium tuberculosis H37Rv            H37Rv     taxon:83332   \n",
       "20     Saccharophagus degradans 2-40             2-40    taxon:203122   \n",
       "21    Sphingopyxis alaskensis RB2256           RB2256    taxon:317655   \n",
       "27    Yersinia mollaretii ATCC 43969       ATCC 43969    taxon:349967   \n",
       "30          Serratia marcescens W2.3             W2.3   taxon:1218513   \n",
       "\n",
       "                              strain  \\\n",
       "0     Pseudomonas aeruginosa PA38182   \n",
       "2            Paenibacillus sp. LC231   \n",
       "4   Mycobacterium tuberculosis H37Rv   \n",
       "20     Saccharophagus degradans 2-40   \n",
       "21    Sphingopyxis alaskensis RB2256   \n",
       "27    Yersinia mollaretii ATCC 43969   \n",
       "30          Serratia marcescens W2.3   \n",
       "\n",
       "                                parent_taxon  strain_wd_id  \n",
       "0      http://www.wikidata.org/entity/Q31856           123  \n",
       "2   http://www.wikidata.org/entity/Q26270468           123  \n",
       "4     http://www.wikidata.org/entity/Q130971           123  \n",
       "20   http://www.wikidata.org/entity/Q7396606           123  \n",
       "21  http://www.wikidata.org/entity/Q21324563           123  \n",
       "27  http://www.wikidata.org/entity/Q16994539           123  \n",
       "30    http://www.wikidata.org/entity/Q140004           123  \n",
       "\n",
       "[7 rows x 41 columns]"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##################\n",
    "### DISCLAIMER ###\n",
    "##################\n",
    "# This is not finished yet - I have to dive deeper into this \n",
    "# will be finished in the coming days \n",
    "\n",
    "z = df.iloc[0:7].copy()\n",
    "z[\"strain_wd_id\"] = z.apply(wikidata_wrapper, axis=1)\n",
    "z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "4d8f46ab-9e44-4769-940c-8f3b9b30a46d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCBI Taxonomy ID 1263833 added to Q124664471.\n"
     ]
    }
   ],
   "source": [
    "def add_ncbi_taxonomy_id(item_id, ncbi_taxonomy_id):\n",
    "    site = pywikibot.Site(\"wikidata\", \"wikidata\")\n",
    "    repo = site.data_repository()\n",
    "\n",
    "    # Load the Wikidata item\n",
    "    item = pywikibot.ItemPage(repo, item_id)\n",
    "    item.get()\n",
    "\n",
    "    # Check if the identifier is already present\n",
    "    if \"P685\" in item.claims:\n",
    "        existing_identifiers = [claim.target.amount for claim in item.claims[\"P685\"]]\n",
    "        if ncbi_taxonomy_id in existing_identifiers:\n",
    "            print(f\"NCBI Taxonomy ID {ncbi_taxonomy_id} already exists for {item_id}.\")\n",
    "            return\n",
    "\n",
    "    # Add the identifier to the item\n",
    "    new_claim = pywikibot.Claim(repo, \"P685\")\n",
    "    new_claim.setTarget(ncbi_taxonomy_id)\n",
    "    item.addClaim(new_claim)\n",
    "\n",
    "    print(f\"NCBI Taxonomy ID {ncbi_taxonomy_id} added to {item_id}.\")\n",
    "\n",
    "# Example usage:\n",
    "wikidata_item_id = \"Q124664471\"  # Replace with the Wikidata item ID you're working with\n",
    "ncbi_taxonomy_id_to_add = \"1263833\"  # Replace with the NCBI Taxonomy ID you want to add\n",
    "\n",
    "add_ncbi_taxonomy_id(wikidata_item_id, ncbi_taxonomy_id_to_add)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "ded584d7-f1ba-453a-a7f6-2020d8dbbca4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Q124664471'"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site = pywikibot.Site(\"test\", \"wikidata\")\n",
    "repo = site.data_repository()\n",
    "item = pywikibot.ItemPage(repo, \"Q124664471\")\n",
    "item.getID()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "625139ca-bdd9-414e-99d6-304ce4c897ac",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'on_item'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[156], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m claim \u001b[38;5;241m=\u001b[39m pywikibot\u001b[38;5;241m.\u001b[39mClaim(repo, \u001b[38;5;124mu\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mP685\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# target = pywikibot.ItemPage(repo, )\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# claim.setTarget(u\"1263833\")\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[43mclaim\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43maddQualifier\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mu\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m1263833\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msummary\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43mu\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mAdding a qualifier.\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/python-venvs/lodkg/lib/python3.10/site-packages/pywikibot/page/_wikibase.py:1964\u001b[0m, in \u001b[0;36mClaim.addQualifier\u001b[0;34m(self, qualifier, **kwargs)\u001b[0m\n\u001b[1;32m   1958\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Add the given qualifier.\u001b[39;00m\n\u001b[1;32m   1959\u001b[0m \n\u001b[1;32m   1960\u001b[0m \u001b[38;5;124;03m:param qualifier: the qualifier to add\u001b[39;00m\n\u001b[1;32m   1961\u001b[0m \u001b[38;5;124;03m:type qualifier: pywikibot.page.Claim\u001b[39;00m\n\u001b[1;32m   1962\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1963\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_assert_mainsnak(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCannot add qualifiers to a \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m-> 1964\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mqualifier\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mon_item\u001b[49m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1965\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1966\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThe provided Claim instance is already used in an entity\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m   1967\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mon_item \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'on_item'"
     ]
    }
   ],
   "source": [
    "\n",
    "qualifier = pywikibot.Claim(repo, u'P685')\n",
    "target = pywikibot.ItemPage(repo, \"Q35409\")\n",
    "qualifier.setTarget(target)\n",
    "claim.addQualifier(qualifier, summary=u'Adding a qualifier.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "e7644c81-6ead-4062-935e-c9a54f9e22b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DATA_ATTRIBUTES',\n",
       " '__abstractmethods__',\n",
       " '__annotations__',\n",
       " '__class__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattr__',\n",
       " '__getattribute__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__slots__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_abc_impl',\n",
       " '_cache_attrs',\n",
       " '_check_bot_may_edit',\n",
       " '_cmpkey',\n",
       " '_cosmetic_changes_hook',\n",
       " '_defined_by',\n",
       " '_getInternals',\n",
       " '_initialize_empty',\n",
       " '_latest_cached_revision',\n",
       " '_link',\n",
       " '_namespace',\n",
       " '_normalizeData',\n",
       " '_revisions',\n",
       " '_save',\n",
       " 'addClaim',\n",
       " 'applicable_protections',\n",
       " 'autoFormat',\n",
       " 'backlinks',\n",
       " 'botMayEdit',\n",
       " 'categories',\n",
       " 'change_category',\n",
       " 'clear_cache',\n",
       " 'concept_uri',\n",
       " 'content_model',\n",
       " 'contributors',\n",
       " 'coordinates',\n",
       " 'create_short_link',\n",
       " 'data_item',\n",
       " 'data_repository',\n",
       " 'defaultsort',\n",
       " 'delete',\n",
       " 'depth',\n",
       " 'editAliases',\n",
       " 'editDescriptions',\n",
       " 'editEntity',\n",
       " 'editLabels',\n",
       " 'editTime',\n",
       " 'embeddedin',\n",
       " 'entity_type',\n",
       " 'exists',\n",
       " 'expand_text',\n",
       " 'extlinks',\n",
       " 'extract',\n",
       " 'fromPage',\n",
       " 'from_entity_uri',\n",
       " 'full_url',\n",
       " 'get',\n",
       " 'getCategoryRedirectTarget',\n",
       " 'getDeletedRevision',\n",
       " 'getID',\n",
       " 'getOldVersion',\n",
       " 'getRedirectTarget',\n",
       " 'getReferences',\n",
       " 'getSitelink',\n",
       " 'getVersionHistoryTable',\n",
       " 'get_data_for_new_entity',\n",
       " 'get_parsed_page',\n",
       " 'has_content',\n",
       " 'has_deleted_revisions',\n",
       " 'has_permission',\n",
       " 'id',\n",
       " 'image_repository',\n",
       " 'imagelinks',\n",
       " 'interwiki',\n",
       " 'isAutoTitle',\n",
       " 'isCategoryRedirect',\n",
       " 'isDisambig',\n",
       " 'isIpEdit',\n",
       " 'isRedirectPage',\n",
       " 'isStaticRedirect',\n",
       " 'isTalkPage',\n",
       " 'is_categorypage',\n",
       " 'is_filepage',\n",
       " 'is_flow_page',\n",
       " 'is_valid_id',\n",
       " 'iterlanglinks',\n",
       " 'iterlinks',\n",
       " 'itertemplates',\n",
       " 'langlinks',\n",
       " 'lastNonBotUser',\n",
       " 'latest_revision',\n",
       " 'latest_revision_id',\n",
       " 'linkedPages',\n",
       " 'loadDeletedRevisions',\n",
       " 'markDeletedRevision',\n",
       " 'mergeInto',\n",
       " 'merge_history',\n",
       " 'move',\n",
       " 'moved_target',\n",
       " 'namespace',\n",
       " 'oldest_revision',\n",
       " 'page_image',\n",
       " 'pageid',\n",
       " 'permalink',\n",
       " 'preloadText',\n",
       " 'properties',\n",
       " 'protect',\n",
       " 'protection',\n",
       " 'purge',\n",
       " 'put',\n",
       " 'redirects',\n",
       " 'removeClaims',\n",
       " 'removeSitelink',\n",
       " 'removeSitelinks',\n",
       " 'repo',\n",
       " 'revision_count',\n",
       " 'revisions',\n",
       " 'save',\n",
       " 'section',\n",
       " 'setSitelink',\n",
       " 'setSitelinks',\n",
       " 'set_redirect_target',\n",
       " 'site',\n",
       " 'templates',\n",
       " 'text',\n",
       " 'title',\n",
       " 'title_pattern',\n",
       " 'toJSON',\n",
       " 'toggleTalkPage',\n",
       " 'touch',\n",
       " 'undelete',\n",
       " 'userName',\n",
       " 'version',\n",
       " 'watch']"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5beacb1-9f62-48e1-91ee-23f6c08429aa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "154e3f49-8b81-4d1b-8e77-c2f6e1df5101",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef6a879a-c3fc-4c51-b383-daa20fe978e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iloc[0]: \n",
    "# iloc[1]: \n",
    "# iloc[2]: \n",
    "# iloc[3]: already exists\n",
    "# iloc[4]: Q124664441\n",
    "# iloc[5]: Q124664454\n",
    "# iloc[6]: Q124664466\n",
    "# iloc[7]: Q124664471"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "6d415dc5-8ad6-4fc0-a036-8c85547421b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organism2</th>\n",
       "      <th>parent_taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>http://www.wikidata.org/entity/Q31856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>http://www.wikidata.org/entity/Q26270468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycobacterium tuberculosis H37Rv</td>\n",
       "      <td>http://www.wikidata.org/entity/Q130971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Saccharophagus degradans 2-40</td>\n",
       "      <td>http://www.wikidata.org/entity/Q7396606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Sphingopyxis alaskensis RB2256</td>\n",
       "      <td>http://www.wikidata.org/entity/Q21324563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Yersinia mollaretii ATCC 43969</td>\n",
       "      <td>http://www.wikidata.org/entity/Q16994539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Serratia marcescens W2.3</td>\n",
       "      <td>http://www.wikidata.org/entity/Q140004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   genbank_organism2                              parent_taxon\n",
       "0     Pseudomonas aeruginosa PA38182     http://www.wikidata.org/entity/Q31856\n",
       "2            Paenibacillus sp. LC231  http://www.wikidata.org/entity/Q26270468\n",
       "4   Mycobacterium tuberculosis H37Rv    http://www.wikidata.org/entity/Q130971\n",
       "20     Saccharophagus degradans 2-40   http://www.wikidata.org/entity/Q7396606\n",
       "21    Sphingopyxis alaskensis RB2256  http://www.wikidata.org/entity/Q21324563\n",
       "27    Yersinia mollaretii ATCC 43969  http://www.wikidata.org/entity/Q16994539\n",
       "30          Serratia marcescens W2.3    http://www.wikidata.org/entity/Q140004"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(7)[[\"genbank_organism2\", \"parent_taxon\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b32a66cd-c2d4-42c2-b50e-300f0a4c5f0f",
   "metadata": {},
   "source": [
    "## Discussion and Conclusion\n",
    "\n",
    "In conclusion, it can be noted that data has been successfully queried from two databases (GenBank and RefSeq). However, due to incomplete data in some cases (references to higher taxons, the use of abbreviations leading to ambiguity in bacterial species, or discrepancies between GenBank and RefSeq in providing different names), the originally extensive dataset of approximately 10,000 entries has been reduced to a more manageable 300 entries that can be reliably implemented.\n",
    "\n",
    "Following the model of other bacteria, genes, and proteins (with the addition of the tag \"response to antibiotic\" here), these selected entries are implemented. However, unlike other genes already present in Wikidata, the \"Entrez Gene ID\" reference could not be found. Despite this missing link, the genes are implemented in Wikidata with the hope that users will contribute and add this information in the future.\n",
    "\n",
    "In summary, it can be concluded that a small percentage of all antibiotic-resistant bacteria known to NCBI could potentially be implemented in Wikidata. The main obstacles to implementing further data are the incomplete or non-standardized entries in GenBank and RefSeq databases and the absence of expert knowledge."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2158afc1-d713-4ee7-a55b-7c992286fab5",
   "metadata": {},
   "source": [
    "## Link to GitHub\n",
    "\n",
    "\n",
    "https://github.com/gjmm07/DS_LOD_and_Knowledge_Graphs_2023_Finn_Heydemann"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32701d57-d5fa-4f7a-8837-623c87b2ae07",
   "metadata": {},
   "source": [
    "## Litarture"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6995775-04f2-4d45-b2d6-77e22aa85264",
   "metadata": {},
   "source": [
    "[1] Salam et al.: Antimicrobial Resistance: A Growing Serious Threat for Global Public Health, 2023\n",
    "\n",
    "[2] National Center for Biotechnolog Information: National Database of Antibiotic Resistant Organisms (NDARO), URL: https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/, last accessed: 26.02.2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91a2bec0-7732-4186-83dc-1ffe7faee410",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lodkg",
   "language": "python",
   "name": "lodkg"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
