{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6442f700-4975-4dd2-b82b-792c6472634a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import urllib\n",
    "import time \n",
    "import numpy as np \n",
    "import bioservice_fetcher as biof \n",
    "import os "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e328c2b-56f0-43b8-9211-3856cd73464e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Fetch data from NCBI database \n",
    "\n",
    "The read dataframe contains keys to access the protein, nucleotide either via Reference Sequence (RefSeq) database or genbank. Lets load the first two columns of the dataframe and read some values. First try is to get the data from the RefSeq-database.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4f4fe93-91d3-48bb-9a65-626aea9a69cf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele</th>\n",
       "      <th>gene_family</th>\n",
       "      <th>whitelisted_taxa</th>\n",
       "      <th>product_name</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>subtype</th>\n",
       "      <th>class</th>\n",
       "      <th>subclass</th>\n",
       "      <th>refseq_protein_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>refseq_protein</th>\n",
       "      <th>refseq_genome</th>\n",
       "      <th>refseq_organism</th>\n",
       "      <th>refseq_parent</th>\n",
       "      <th>rrefseq_protein</th>\n",
       "      <th>refseq_same_parent</th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "      <th>genbank_strain_nuc</th>\n",
       "      <th>genbank_organsim_prot</th>\n",
       "      <th>genbank_strain_prot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4666</th>\n",
       "      <td>blaPLA-4</td>\n",
       "      <td>blaPLA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>class A beta-lactamase PLA-4</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>WP_032687534.1</td>\n",
       "      <td>...</td>\n",
       "      <td>class A beta-lactamase PLA-4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Raoultella planticola</td>\n",
       "      <td>Raoultella planticola</td>\n",
       "      <td>class A beta-lactamase PLA-4</td>\n",
       "      <td>Raoultella planticola</td>\n",
       "      <td>Raoultella planticola</td>\n",
       "      <td>42-1</td>\n",
       "      <td>Raoultella planticola</td>\n",
       "      <td>42-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5397</th>\n",
       "      <td>catB6</td>\n",
       "      <td>catB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>type B-3 chloramphenicol O-acetyltransferase C...</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>PHENICOL</td>\n",
       "      <td>CHLORAMPHENICOL</td>\n",
       "      <td>WP_063843229.1</td>\n",
       "      <td>...</td>\n",
       "      <td>type B-3 chloramphenicol O-acetyltransferase C...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>type B-3 chloramphenicol O-acetyltransferase C...</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8260</th>\n",
       "      <td>fusA_G452S</td>\n",
       "      <td>fusA</td>\n",
       "      <td>Staphylococcus_aureus</td>\n",
       "      <td>elongation factor G</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>POINT</td>\n",
       "      <td>FUSIDIC ACID</td>\n",
       "      <td>FUSIDIC ACID</td>\n",
       "      <td>WP_000090315.1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>elongation factor G</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus Mu50</td>\n",
       "      <td>Mu50</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus Mu50</td>\n",
       "      <td>Mu50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>blaFLC-1</td>\n",
       "      <td>blaFRI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FRI family carbapenem-hydrolyzing class A beta...</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>BETA-LACTAM</td>\n",
       "      <td>CARBAPENEM</td>\n",
       "      <td>WP_123061077.1</td>\n",
       "      <td>...</td>\n",
       "      <td>FRI family carbapenem-hydrolyzing class A beta...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacter cloacae</td>\n",
       "      <td>Enterobacter cloacae</td>\n",
       "      <td>FRI family carbapenem-hydrolyzing class A beta...</td>\n",
       "      <td>Enterobacter cloacae</td>\n",
       "      <td>Enterobacter cloacae</td>\n",
       "      <td>FRI-3442</td>\n",
       "      <td>Enterobacter cloacae</td>\n",
       "      <td>FRI-3442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7825</th>\n",
       "      <td>NaN</td>\n",
       "      <td>vanXY-N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D-Ala-D-Ala dipeptidase/D-Ala-D-Ala carboxypep...</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "      <td>AMR</td>\n",
       "      <td>GLYCOPEPTIDE</td>\n",
       "      <td>VANCOMYCIN</td>\n",
       "      <td>WP_063856820.1</td>\n",
       "      <td>...</td>\n",
       "      <td>D-Ala-D-Ala dipeptidase/D-Ala-D-Ala carboxypep...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>Enterococcus</td>\n",
       "      <td>D-Ala-D-Ala dipeptidase/D-Ala-D-Ala carboxypep...</td>\n",
       "      <td>Enterococcus</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>UCN71</td>\n",
       "      <td>Enterococcus faecium</td>\n",
       "      <td>UCN71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          allele gene_family       whitelisted_taxa  \\\n",
       "4666    blaPLA-4      blaPLA                    NaN   \n",
       "5397       catB6        catB                    NaN   \n",
       "8260  fusA_G452S        fusA  Staphylococcus_aureus   \n",
       "1990    blaFLC-1      blaFRI                    NaN   \n",
       "7825         NaN     vanXY-N                    NaN   \n",
       "\n",
       "                                           product_name scope type subtype  \\\n",
       "4666                       class A beta-lactamase PLA-4  core  AMR     AMR   \n",
       "5397  type B-3 chloramphenicol O-acetyltransferase C...  core  AMR     AMR   \n",
       "8260                                elongation factor G  core  AMR   POINT   \n",
       "1990  FRI family carbapenem-hydrolyzing class A beta...  core  AMR     AMR   \n",
       "7825  D-Ala-D-Ala dipeptidase/D-Ala-D-Ala carboxypep...  core  AMR     AMR   \n",
       "\n",
       "             class         subclass refseq_protein_accession  ...  \\\n",
       "4666   BETA-LACTAM      BETA-LACTAM           WP_032687534.1  ...   \n",
       "5397      PHENICOL  CHLORAMPHENICOL           WP_063843229.1  ...   \n",
       "8260  FUSIDIC ACID     FUSIDIC ACID           WP_000090315.1  ...   \n",
       "1990   BETA-LACTAM       CARBAPENEM           WP_123061077.1  ...   \n",
       "7825  GLYCOPEPTIDE       VANCOMYCIN           WP_063856820.1  ...   \n",
       "\n",
       "                                         refseq_protein refseq_genome  \\\n",
       "4666                       class A beta-lactamase PLA-4           NaN   \n",
       "5397  type B-3 chloramphenicol O-acetyltransferase C...           NaN   \n",
       "8260                                                NaN           NaN   \n",
       "1990  FRI family carbapenem-hydrolyzing class A beta...           NaN   \n",
       "7825  D-Ala-D-Ala dipeptidase/D-Ala-D-Ala carboxypep...           NaN   \n",
       "\n",
       "             refseq_organism          refseq_parent  \\\n",
       "4666   Raoultella planticola  Raoultella planticola   \n",
       "5397  Pseudomonas aeruginosa    Gammaproteobacteria   \n",
       "8260                     NaN         Staphylococcus   \n",
       "1990    Enterobacter cloacae   Enterobacter cloacae   \n",
       "7825    Enterococcus faecium           Enterococcus   \n",
       "\n",
       "                                        rrefseq_protein  \\\n",
       "4666                       class A beta-lactamase PLA-4   \n",
       "5397  type B-3 chloramphenicol O-acetyltransferase C...   \n",
       "8260                                elongation factor G   \n",
       "1990  FRI family carbapenem-hydrolyzing class A beta...   \n",
       "7825  D-Ala-D-Ala dipeptidase/D-Ala-D-Ala carboxypep...   \n",
       "\n",
       "         refseq_same_parent                      genbank_organsim_nuc  \\\n",
       "4666  Raoultella planticola                     Raoultella planticola   \n",
       "5397    Gammaproteobacteria                    Pseudomonas aeruginosa   \n",
       "8260         Staphylococcus  Staphylococcus aureus subsp. aureus Mu50   \n",
       "1990   Enterobacter cloacae                      Enterobacter cloacae   \n",
       "7825           Enterococcus                      Enterococcus faecium   \n",
       "\n",
       "     genbank_strain_nuc                     genbank_organsim_prot  \\\n",
       "4666               42-1                     Raoultella planticola   \n",
       "5397                NaN                    Pseudomonas aeruginosa   \n",
       "8260               Mu50  Staphylococcus aureus subsp. aureus Mu50   \n",
       "1990           FRI-3442                      Enterobacter cloacae   \n",
       "7825              UCN71                      Enterococcus faecium   \n",
       "\n",
       "      genbank_strain_prot  \n",
       "4666                 42-1  \n",
       "5397                  NaN  \n",
       "8260                 Mu50  \n",
       "1990             FRI-3442  \n",
       "7825                UCN71  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reads everything that could possible be interesting for this project\n",
    "\n",
    "def fetch_data(read_from_web: bool = False) -> pd.DataFrame: \n",
    "    \"\"\"\n",
    "    if fetch_data_switch is True data will be fetched from NCBI and interesting things will be read from genbank or refseq database -- caution: Takes for ages \n",
    "    else data will be read from last time -- should be used in most of the cases \n",
    "    \"\"\"\n",
    "    if read_from_web: \n",
    "        url = \"https://ftp.ncbi.nlm.nih.gov/pathogen/Antimicrobial_resistance/AMRFinderPlus/database/latest/ReferenceGeneCatalog.txt\"\n",
    "        df = pd.read_csv(urllib.request.urlopen(url), delimiter=\"\\t\")\n",
    "        df = df.sample(10)\n",
    "        df.apply(biof.get_protein_and_parent, axis=1)\n",
    "        df.apply(biof.get_strain_and_gene, axis=1)\n",
    "        df.apply(biof.get_organism_strain_via_prot,axis=1)\n",
    "        df.apply(biof.get_organism_strain_via_nuc, axis=1)\n",
    "        df.to_csv(\"resistance_df.csv\", index=False)\n",
    "    else: \n",
    "        if not os.path.exists(\"resistance_df.csv\"): \n",
    "            print(\"Cannot read from hard drive because file does not exist -- set read from web switch to True\")\n",
    "            return None\n",
    "        df = pd.read_csv(\"resistance_df.csv\")\n",
    "    return df\n",
    "\n",
    "\n",
    "df = fetch_data()\n",
    "df.sample(5, random_state=19)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffd50fed-0478-4c5e-86f5-77b4d268ea51",
   "metadata": {
    "tags": []
   },
   "source": [
    "## STRAIN\n",
    "\n",
    "Because there is quite a bit of data present, my goal is to extract the right name of the becterial strain. Therefore I am going to find the best datasource of organsim name and combine it with the best source of the exact strain name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "16a7848f-4bdd-4011-837e-dcd9f3c1adf0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['allele', 'gene_family', 'whitelisted_taxa', 'product_name', 'scope',\n",
       "       'type', 'subtype', 'class', 'subclass', 'refseq_protein_accession',\n",
       "       'refseq_nucleotide_accession', 'curated_refseq_start',\n",
       "       'genbank_protein_accession', 'genbank_nucleotide_accession',\n",
       "       'genbank_strand', 'genbank_start', 'genbank_stop', 'refseq_strand',\n",
       "       'refseq_start', 'refseq_stop', 'pubmed_reference', 'blacklisted_taxa',\n",
       "       'synonyms', 'hierarchy_node', 'db_version', 'refseq_gene',\n",
       "       'refseq_protein', 'refseq_genome', 'refseq_organism', 'refseq_parent',\n",
       "       'rrefseq_protein', 'refseq_same_parent', 'genbank_organsim_nuc',\n",
       "       'genbank_strain_nuc', 'genbank_organsim_prot', 'genbank_strain_prot'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e2217a5c-536c-46c0-ad9b-8aa04f75f4de",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq_organism</th>\n",
       "      <th>refseq_parent</th>\n",
       "      <th>refseq_same_parent</th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "      <th>genbank_organsim_prot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>Simplicispira metamorpha</td>\n",
       "      <td>Simplicispira metamorpha</td>\n",
       "      <td>Simplicispira metamorpha</td>\n",
       "      <td>Simplicispira metamorpha</td>\n",
       "      <td>Simplicispira metamorpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2293</th>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae subsp. pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6959</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8684</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144</th>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6553</th>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3522</th>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7193</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia phage 933W</td>\n",
       "      <td>Escherichia phage 933W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4930</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              refseq_organism             refseq_parent  \\\n",
       "389                  Simplicispira metamorpha  Simplicispira metamorpha   \n",
       "2293  Klebsiella pneumoniae subsp. pneumoniae     Klebsiella pneumoniae   \n",
       "6959                         Escherichia coli       Gammaproteobacteria   \n",
       "8684                                      NaN                Klebsiella   \n",
       "4144                   Pseudomonas aeruginosa    Pseudomonas aeruginosa   \n",
       "1563                    Klebsiella pneumoniae        Enterobacteriaceae   \n",
       "6553                  Acinetobacter baumannii                  Bacteria   \n",
       "3522                     Campylobacter jejuni      Campylobacter jejuni   \n",
       "7193                                      NaN                       NaN   \n",
       "4930                         Escherichia coli          Escherichia coli   \n",
       "\n",
       "            refseq_same_parent                     genbank_organsim_nuc  \\\n",
       "389   Simplicispira metamorpha                 Simplicispira metamorpha   \n",
       "2293     Klebsiella pneumoniae  Klebsiella pneumoniae subsp. pneumoniae   \n",
       "6959       Gammaproteobacteria                         Escherichia coli   \n",
       "8684                Klebsiella                    Klebsiella pneumoniae   \n",
       "4144    Pseudomonas aeruginosa                   Pseudomonas aeruginosa   \n",
       "1563        Enterobacteriaceae                    Klebsiella pneumoniae   \n",
       "6553                  Bacteria                  Acinetobacter baumannii   \n",
       "3522      Campylobacter jejuni                     Campylobacter jejuni   \n",
       "7193                       NaN                   Escherichia phage 933W   \n",
       "4930          Escherichia coli                         Escherichia coli   \n",
       "\n",
       "                        genbank_organsim_prot  \n",
       "389                  Simplicispira metamorpha  \n",
       "2293  Klebsiella pneumoniae subsp. pneumoniae  \n",
       "6959                         Escherichia coli  \n",
       "8684                    Klebsiella pneumoniae  \n",
       "4144                   Pseudomonas aeruginosa  \n",
       "1563                    Klebsiella pneumoniae  \n",
       "6553                  Acinetobacter baumannii  \n",
       "3522                     Campylobacter jejuni  \n",
       "7193                   Escherichia phage 933W  \n",
       "4930                         Escherichia coli  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Just by looking into this small random sampled dataframe \"refseq organism\", \"genbank_organsim_nuc\" and \"genbank_organism_prot\" yield pretty similar results, \n",
    "# altough \"refseq_organism\" has less information\n",
    "# \"refseq_parent\" and \"refseq_same_parent\" sometimes carry the same organism name, but sometimes some higher taxon\n",
    "# For extracting the right organism name I am going to look closer into \"refseq_organism\", \"genbank_organism_nuc\" and \"ganbank_organism_prot\"\n",
    "df[[\"refseq_organism\", \"refseq_parent\", \"refseq_same_parent\", \"genbank_organsim_nuc\", \"genbank_organsim_prot\"]].sample(10, random_state=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2008f299-d7bd-4625-a798-efe9f17982eb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organsim_prot</th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "      <th>refseq_organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Burkholderia glumae</td>\n",
       "      <td>Burkholderia glumae</td>\n",
       "      <td>Burkholderia glumae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Providencia stuartii</td>\n",
       "      <td>Providencia stuartii</td>\n",
       "      <td>Providencia stuartii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mycolicibacterium fortuitum</td>\n",
       "      <td>Mycolicibacterium fortuitum</td>\n",
       "      <td>Mycolicibacterium fortuitum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9184</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9185</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9186</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9187</th>\n",
       "      <td>Burkholderia mallei ATCC 23344</td>\n",
       "      <td>Burkholderia mallei ATCC 23344</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9188</th>\n",
       "      <td>Enterococcus faecalis V583</td>\n",
       "      <td>Enterococcus faecalis V583</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9189 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            genbank_organsim_prot  \\\n",
       "0                  Pseudomonas aeruginosa PA38182   \n",
       "1                             Burkholderia glumae   \n",
       "2                         Paenibacillus sp. LC231   \n",
       "3                            Providencia stuartii   \n",
       "4                     Mycolicibacterium fortuitum   \n",
       "...                                           ...   \n",
       "9184  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9185  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9186  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9187               Burkholderia mallei ATCC 23344   \n",
       "9188                   Enterococcus faecalis V583   \n",
       "\n",
       "                             genbank_organsim_nuc  \\\n",
       "0                  Pseudomonas aeruginosa PA38182   \n",
       "1                             Burkholderia glumae   \n",
       "2                         Paenibacillus sp. LC231   \n",
       "3                            Providencia stuartii   \n",
       "4                     Mycolicibacterium fortuitum   \n",
       "...                                           ...   \n",
       "9184  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9185  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9186  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9187               Burkholderia mallei ATCC 23344   \n",
       "9188                   Enterococcus faecalis V583   \n",
       "\n",
       "                     refseq_organism  \n",
       "0     Pseudomonas aeruginosa PA38182  \n",
       "1                Burkholderia glumae  \n",
       "2            Paenibacillus sp. LC231  \n",
       "3               Providencia stuartii  \n",
       "4        Mycolicibacterium fortuitum  \n",
       "...                              ...  \n",
       "9184                             NaN  \n",
       "9185                             NaN  \n",
       "9186                             NaN  \n",
       "9187                             NaN  \n",
       "9188                             NaN  \n",
       "\n",
       "[9189 rows x 3 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[\"genbank_organsim_prot\", \"genbank_organsim_nuc\", \"refseq_organism\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "c34bff8f-1f7e-48ba-9cff-178e41936a0d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organsim_prot</th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "      <th>refseq_organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Serratia marcescens</td>\n",
       "      <td>Serratia marcescens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NaN</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "      <td>uncultured bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Paracoccus denitrificans PD1222</td>\n",
       "      <td>Paracoccus denitrificans PD1222</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli APEC O1</td>\n",
       "      <td>Escherichia coli APEC O1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptomyces xiaopingdaonensis</td>\n",
       "      <td>Streptomyces xiaopingdaonensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9184</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9185</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9186</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9187</th>\n",
       "      <td>Burkholderia mallei ATCC 23344</td>\n",
       "      <td>Burkholderia mallei ATCC 23344</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9188</th>\n",
       "      <td>Enterococcus faecalis V583</td>\n",
       "      <td>Enterococcus faecalis V583</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2651 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            genbank_organsim_prot  \\\n",
       "34                                            NaN   \n",
       "36                                            NaN   \n",
       "45                Paracoccus denitrificans PD1222   \n",
       "52                                            NaN   \n",
       "58                                            NaN   \n",
       "...                                           ...   \n",
       "9184  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9185  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9186  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9187               Burkholderia mallei ATCC 23344   \n",
       "9188                   Enterococcus faecalis V583   \n",
       "\n",
       "                             genbank_organsim_nuc  \\\n",
       "34                            Serratia marcescens   \n",
       "36                           uncultured bacterium   \n",
       "45                Paracoccus denitrificans PD1222   \n",
       "52                       Escherichia coli APEC O1   \n",
       "58                 Streptomyces xiaopingdaonensis   \n",
       "...                                           ...   \n",
       "9184  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9185  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9186  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9187               Burkholderia mallei ATCC 23344   \n",
       "9188                   Enterococcus faecalis V583   \n",
       "\n",
       "                     refseq_organism  \n",
       "34               Serratia marcescens  \n",
       "36              uncultured bacterium  \n",
       "45                               NaN  \n",
       "52          Escherichia coli APEC O1  \n",
       "58    Streptomyces xiaopingdaonensis  \n",
       "...                              ...  \n",
       "9184                             NaN  \n",
       "9185                             NaN  \n",
       "9186                             NaN  \n",
       "9187                             NaN  \n",
       "9188                             NaN  \n",
       "\n",
       "[2651 rows x 3 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2651 cases where found organisms in genbank via protein, genbank via nuclotide and refseq do not match \n",
    "\n",
    "diff_organism_df = df.loc[(df[\"genbank_organsim_nuc\"] != df[\"genbank_organsim_prot\"]) | (df[\"genbank_organsim_nuc\"] != df[\"refseq_organism\"]), [\"genbank_organsim_prot\", \n",
    "                                                                                                                                                \"genbank_organsim_nuc\", \n",
    "                                                                                                                                                \"refseq_organism\"]]\n",
    "diff_organism_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b62f37a1-67da-462a-9917-b3143af6daca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genbank_organsim_prot     363\n",
       "genbank_organsim_nuc        0\n",
       "refseq_organism          2396\n",
       "dtype: int64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# But genbank via nucleotid has non NaN -- genbank via nucleotid seems to be best: have a look at the exact differences \n",
    "diff_organism_df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "30f6564d-24b1-47ee-ab09-3b437ebafc95",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "      <th>genbank_organsim_prot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5382</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5600</th>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5612</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5833</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6103</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6106</th>\n",
       "      <td>Staphylococcus epidermidis RP62A phage SP-beta</td>\n",
       "      <td>Staphylococcus epidermidis RP62A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6109</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus phage lambda Ba03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6145</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6146</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6151</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6152</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6153</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6163</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6193</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6194</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6429</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6430</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6611</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6755</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6756</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6991</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7053</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7061</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Phage Gifsy-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7302</th>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7303</th>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7304</th>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7305</th>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7306</th>\n",
       "      <td>Peptoclostridium phage p630P2</td>\n",
       "      <td>Clostridioides difficile 630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7911</th>\n",
       "      <td>Streptococcus phage PhiSpn_200</td>\n",
       "      <td>Streptococcus pneumoniae AP200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7920</th>\n",
       "      <td>Streptococcus phage PhiSpn_200</td>\n",
       "      <td>Streptococcus pneumoniae AP200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7926</th>\n",
       "      <td>Streptococcus phage PhiSpn_200</td>\n",
       "      <td>Streptococcus pneumoniae AP200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9156</th>\n",
       "      <td>Salmonella virus Fels2</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                genbank_organsim_nuc  \\\n",
       "126                           Salmonella virus Fels2   \n",
       "599                       Bacillus phage lambda Ba02   \n",
       "5382                                Escherichia coli   \n",
       "5600                   Peptoclostridium phage p630P2   \n",
       "5612                         Bacillus phage phBC6A52   \n",
       "5833                         Bacillus phage phBC6A52   \n",
       "6103                         Bacillus phage phBC6A52   \n",
       "6106  Staphylococcus epidermidis RP62A phage SP-beta   \n",
       "6109                      Bacillus phage lambda Ba02   \n",
       "6145                          Salmonella virus Fels2   \n",
       "6146                          Salmonella virus Fels2   \n",
       "6151                         Bacillus phage phBC6A52   \n",
       "6152                         Bacillus phage phBC6A52   \n",
       "6153                         Bacillus phage phBC6A52   \n",
       "6163                         Bacillus phage phBC6A52   \n",
       "6193                         Bacillus phage phBC6A52   \n",
       "6194                      Bacillus phage lambda Ba02   \n",
       "6429                          Salmonella virus Fels2   \n",
       "6430                          Salmonella virus Fels2   \n",
       "6611                         Bacillus phage phBC6A52   \n",
       "6755                         Bacillus phage phBC6A52   \n",
       "6756                      Bacillus phage lambda Ba02   \n",
       "6991                      Bacillus phage lambda Ba02   \n",
       "7053                          Salmonella virus Fels2   \n",
       "7061                          Salmonella virus Fels2   \n",
       "7063                      Bacillus phage lambda Ba02   \n",
       "7302                   Peptoclostridium phage p630P2   \n",
       "7303                   Peptoclostridium phage p630P2   \n",
       "7304                   Peptoclostridium phage p630P2   \n",
       "7305                   Peptoclostridium phage p630P2   \n",
       "7306                   Peptoclostridium phage p630P2   \n",
       "7911                  Streptococcus phage PhiSpn_200   \n",
       "7920                  Streptococcus phage PhiSpn_200   \n",
       "7926                  Streptococcus phage PhiSpn_200   \n",
       "9156                          Salmonella virus Fels2   \n",
       "\n",
       "                                  genbank_organsim_prot  \n",
       "126   Salmonella enterica subsp. enterica serovar Ty...  \n",
       "599                        Bacillus anthracis str. Ames  \n",
       "5382                              Escherichia coli K-12  \n",
       "5600                       Clostridioides difficile 630  \n",
       "5612                         Bacillus cereus ATCC 14579  \n",
       "5833                         Bacillus cereus ATCC 14579  \n",
       "6103                         Bacillus cereus ATCC 14579  \n",
       "6106                   Staphylococcus epidermidis RP62A  \n",
       "6109                         Bacillus phage lambda Ba03  \n",
       "6145  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "6146  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "6151                         Bacillus cereus ATCC 14579  \n",
       "6152                         Bacillus cereus ATCC 14579  \n",
       "6153                         Bacillus cereus ATCC 14579  \n",
       "6163                         Bacillus cereus ATCC 14579  \n",
       "6193                         Bacillus cereus ATCC 14579  \n",
       "6194                       Bacillus anthracis str. Ames  \n",
       "6429  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "6430  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "6611                         Bacillus cereus ATCC 14579  \n",
       "6755                         Bacillus cereus ATCC 14579  \n",
       "6756                       Bacillus anthracis str. Ames  \n",
       "6991                       Bacillus anthracis str. Ames  \n",
       "7053  Salmonella enterica subsp. enterica serovar Ty...  \n",
       "7061                                      Phage Gifsy-2  \n",
       "7063                       Bacillus anthracis str. Ames  \n",
       "7302                       Clostridioides difficile 630  \n",
       "7303                       Clostridioides difficile 630  \n",
       "7304                       Clostridioides difficile 630  \n",
       "7305                       Clostridioides difficile 630  \n",
       "7306                       Clostridioides difficile 630  \n",
       "7911                     Streptococcus pneumoniae AP200  \n",
       "7920                     Streptococcus pneumoniae AP200  \n",
       "7926                     Streptococcus pneumoniae AP200  \n",
       "9156  Salmonella enterica subsp. enterica serovar Ty...  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# As can be seen there are 35 issues where organism found via genbank (nucleotid) and via genbank (protein) are different \n",
    "# What should be done here? Are these alternative names -- Expert knowledge requiered \n",
    "diff_organism_df.loc[diff_organism_df[\"genbank_organsim_nuc\"] != diff_organism_df[\"genbank_organsim_prot\"], [\"genbank_organsim_nuc\", \n",
    "                                                                                                             \"genbank_organsim_prot\"]].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "58ac0927-c085-4a51-b489-c5895b707fcc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "      <th>refseq_organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2020</th>\n",
       "      <td>Acinetobacter sp.</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4017</th>\n",
       "      <td>Klebsiella michiganensis</td>\n",
       "      <td>Klebsiella michiganensis M5al</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5010</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5382</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>Escherichia coli K-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6103</th>\n",
       "      <td>Bacillus phage phBC6A52</td>\n",
       "      <td>Bacillus cereus ATCC 14579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6106</th>\n",
       "      <td>Staphylococcus epidermidis RP62A phage SP-beta</td>\n",
       "      <td>Staphylococcus epidermidis RP62A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6109</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6125</th>\n",
       "      <td>Cytobacillus massiliigabonensis</td>\n",
       "      <td>Bacillus massiliigabonensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6991</th>\n",
       "      <td>Bacillus phage lambda Ba02</td>\n",
       "      <td>Bacillus anthracis str. Ames</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                genbank_organsim_nuc  \\\n",
       "2020                               Acinetobacter sp.   \n",
       "4017                        Klebsiella michiganensis   \n",
       "5010                             Salmonella enterica   \n",
       "5382                                Escherichia coli   \n",
       "6103                         Bacillus phage phBC6A52   \n",
       "6106  Staphylococcus epidermidis RP62A phage SP-beta   \n",
       "6109                      Bacillus phage lambda Ba02   \n",
       "6125                 Cytobacillus massiliigabonensis   \n",
       "6991                      Bacillus phage lambda Ba02   \n",
       "\n",
       "                                        refseq_organism  \n",
       "2020                            Acinetobacter baumannii  \n",
       "4017                      Klebsiella michiganensis M5al  \n",
       "5010  Salmonella enterica subsp. enterica serovar In...  \n",
       "5382                              Escherichia coli K-12  \n",
       "6103                         Bacillus cereus ATCC 14579  \n",
       "6106                   Staphylococcus epidermidis RP62A  \n",
       "6109                       Bacillus anthracis str. Ames  \n",
       "6125                        Bacillus massiliigabonensis  \n",
       "6991                       Bacillus anthracis str. Ames  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Furthermore there are 9 issues between genbank via nucleotide and refseq \n",
    "# Are these alternative names? Expert knowledge requiered \n",
    "diff_organism_df.loc[diff_organism_df[\"genbank_organsim_nuc\"] != diff_organism_df[\"refseq_organism\"], [\"genbank_organsim_nuc\", \n",
    "                                                                                                       \"refseq_organism\"]].dropna()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49011782-010e-46c4-83ab-a835a96c04f7",
   "metadata": {
    "tags": []
   },
   "source": [
    "I am going to choose the organism found by genbank via nucleotid as base organism, because it has no NaN and seems to be complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "1345e9e5-0fe3-4b7a-94b6-5674370f2b76",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_strain_nuc</th>\n",
       "      <th>genbank_strain_prot</th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2492</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8003</th>\n",
       "      <td>MS4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7040</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Micromonospora zionensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>M16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8008</th>\n",
       "      <td>SWU02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6030</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3656</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>MC620</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serratia marcescens MC620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>BWH49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7224</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genbank_strain_nuc genbank_strain_prot          genbank_organsim_nuc\n",
       "2492                NaN                 NaN  Stenotrophomonas maltophilia\n",
       "8003                MS4                 NaN         Staphylococcus aureus\n",
       "7040                NaN                 NaN      Micromonospora zionensis\n",
       "300                 M16                 NaN  Stenotrophomonas maltophilia\n",
       "8008              SWU02                 NaN      Streptococcus pneumoniae\n",
       "6030                NaN                 NaN              Escherichia coli\n",
       "3656                NaN                 NaN        Pseudomonas aeruginosa\n",
       "107               MC620                 NaN     Serratia marcescens MC620\n",
       "460               BWH49                 NaN              Escherichia coli\n",
       "7224                NaN                 NaN        Pseudomonas aeruginosa"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Looks like genbank strain nuc has more information but information is sometimes included in genbank_organism_nuc \n",
    "# Serratia marcescens MC620 has MC620 already included\n",
    "df.loc[df[\"genbank_strain_nuc\"] != df[\"genbank_strain_prot\"], [\"genbank_strain_nuc\", \"genbank_strain_prot\", \"genbank_organsim_nuc\"]].sample(10, random_state=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "dad1438d-22a0-45dd-82dc-e49c2b68e3e5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_strain_nuc</th>\n",
       "      <th>genbank_strain_prot</th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>APEC O1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli APEC O1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>DUT 180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptomyces xiaopingdaonensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>W2.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serratia marcescens W2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>DSM 12546</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Saccharospirillum impatiens DSM 12546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>NRRL ISP-5461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptomyces lydicus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8630</th>\n",
       "      <td>NG196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neisseria gonorrhoeae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8631</th>\n",
       "      <td>NG196</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neisseria gonorrhoeae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8675</th>\n",
       "      <td>TUM15753</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neisseria gonorrhoeae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8676</th>\n",
       "      <td>TUM15753</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neisseria gonorrhoeae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8849</th>\n",
       "      <td>CFBR-515</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>317 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genbank_strain_nuc genbank_strain_prot  \\\n",
       "52              APEC O1                 NaN   \n",
       "58              DUT 180                 NaN   \n",
       "59                 W2.3                 NaN   \n",
       "62            DSM 12546                 NaN   \n",
       "87        NRRL ISP-5461                 NaN   \n",
       "...                 ...                 ...   \n",
       "8630              NG196                 NaN   \n",
       "8631              NG196                 NaN   \n",
       "8675           TUM15753                 NaN   \n",
       "8676           TUM15753                 NaN   \n",
       "8849           CFBR-515                 NaN   \n",
       "\n",
       "                       genbank_organsim_nuc  \n",
       "52                 Escherichia coli APEC O1  \n",
       "58           Streptomyces xiaopingdaonensis  \n",
       "59                 Serratia marcescens W2.3  \n",
       "62    Saccharospirillum impatiens DSM 12546  \n",
       "87                     Streptomyces lydicus  \n",
       "...                                     ...  \n",
       "8630                  Neisseria gonorrhoeae  \n",
       "8631                  Neisseria gonorrhoeae  \n",
       "8675                  Neisseria gonorrhoeae  \n",
       "8676                  Neisseria gonorrhoeae  \n",
       "8849                  Staphylococcus aureus  \n",
       "\n",
       "[317 rows x 3 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Genbank strain prot carrys either redundant information or no information at all -- genbank strain via nucleotid will therefore be selected\n",
    "df.loc[(df[\"genbank_strain_nuc\"] != df[\"genbank_strain_prot\"]) & ~df[\"genbank_strain_nuc\"].isna(), [\"genbank_strain_nuc\", \"genbank_strain_prot\", \"genbank_organsim_nuc\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f96c2ab3-5c47-4b2d-8201-46c019530152",
   "metadata": {
    "tags": []
   },
   "source": [
    "To find the correct and full strain name I will combine \"genbank strain via nucleotid\" and \"genbank organism via nucleotid\", only if strain is not included in organism. Else organism itself will be selected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "8c086057-f065-4e2c-96ba-296e1769df8c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genbank_strain_nuc</th>\n",
       "      <th>genbank_organsim_nuc</th>\n",
       "      <th>strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PA38182</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "      <td>Pseudomonas aeruginosa PA38182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5091</td>\n",
       "      <td>Burkholderia glumae</td>\n",
       "      <td>Burkholderia glumae 5091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LC231</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "      <td>Paenibacillus sp. LC231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Providencia stuartii</td>\n",
       "      <td>Providencia stuartii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FC1K</td>\n",
       "      <td>Mycolicibacterium fortuitum</td>\n",
       "      <td>Mycolicibacterium fortuitum FC1K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9184</th>\n",
       "      <td>MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9185</th>\n",
       "      <td>MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9186</th>\n",
       "      <td>MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus MRSA252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9187</th>\n",
       "      <td>ATCC 23344</td>\n",
       "      <td>Burkholderia mallei ATCC 23344</td>\n",
       "      <td>Burkholderia mallei ATCC 23344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9188</th>\n",
       "      <td>V583</td>\n",
       "      <td>Enterococcus faecalis V583</td>\n",
       "      <td>Enterococcus faecalis V583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9189 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genbank_strain_nuc                         genbank_organsim_nuc  \\\n",
       "0               PA38182               Pseudomonas aeruginosa PA38182   \n",
       "1                  5091                          Burkholderia glumae   \n",
       "2                 LC231                      Paenibacillus sp. LC231   \n",
       "3                   NaN                         Providencia stuartii   \n",
       "4                  FC1K                  Mycolicibacterium fortuitum   \n",
       "...                 ...                                          ...   \n",
       "9184            MRSA252  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9185            MRSA252  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9186            MRSA252  Staphylococcus aureus subsp. aureus MRSA252   \n",
       "9187         ATCC 23344               Burkholderia mallei ATCC 23344   \n",
       "9188               V583                   Enterococcus faecalis V583   \n",
       "\n",
       "                                           strain  \n",
       "0                  Pseudomonas aeruginosa PA38182  \n",
       "1                        Burkholderia glumae 5091  \n",
       "2                         Paenibacillus sp. LC231  \n",
       "3                            Providencia stuartii  \n",
       "4                Mycolicibacterium fortuitum FC1K  \n",
       "...                                           ...  \n",
       "9184  Staphylococcus aureus subsp. aureus MRSA252  \n",
       "9185  Staphylococcus aureus subsp. aureus MRSA252  \n",
       "9186  Staphylococcus aureus subsp. aureus MRSA252  \n",
       "9187               Burkholderia mallei ATCC 23344  \n",
       "9188                   Enterococcus faecalis V583  \n",
       "\n",
       "[9189 rows x 3 columns]"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def check_for_combination(df_row: pd.Series) -> bool:\n",
    "    \"\"\"\n",
    "    Checks if genbank_organism and genbank_strain should be connected to one strain\n",
    "    \"\"\"\n",
    "    if not isinstance(df_row[\"genbank_strain_nuc\"], str): \n",
    "        # don't combine organism and strain if is NaN --> Is this correct? \n",
    "        return False\n",
    "    return not (df_row[\"genbank_strain_nuc\"] in df_row[\"genbank_organsim_nuc\"])\n",
    "    \n",
    "    \n",
    "df[\"strain\"] = np.where(df.apply(check_for_combination , axis=1), \n",
    "                           df[\"genbank_organsim_nuc\"] + \" \" + df[\"genbank_strain_nuc\"], \n",
    "                           df[\"genbank_organsim_nuc\"])\n",
    "df[[\"genbank_strain_nuc\", \"genbank_organsim_nuc\", \"strain\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7358438-6e79-4225-88f1-ff9adc7cd87d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "62a84377-9400-406f-b6da-244317337f4d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       Pseudomonas aeruginosa PA38182 aac(2')-I(A267)\n",
       "1                 Burkholderia glumae 5091 aac(2')-IIa\n",
       "2                  Paenibacillus sp. LC231 aac(2')-IIb\n",
       "3                      Providencia stuartii aac(2')-Ia\n",
       "4              Mycobacterium fortuitum FC1K aac(2')-Ib\n",
       "                             ...                      \n",
       "9184                                               NaN\n",
       "9185                                               NaN\n",
       "9186                                               NaN\n",
       "9187                                               NaN\n",
       "9188                                               NaN\n",
       "Name: refseq_gene, Length: 9189, dtype: object"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"refseq_gene\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "296c701f-05b6-4600-a27e-f27dbf9030e0",
   "metadata": {},
   "source": [
    "## Search for connection to Wikidata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "2527c9e3-e9db-4ddf-8dd0-b16d7f24c3d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "\n",
    "sparql = SPARQLWrapper(\"https://query.wikidata.org/sparql\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "792350f6-ad93-4454-a15f-938e16d8c96a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_organism</th>\n",
       "      <th>_strain</th>\n",
       "      <th>parent taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7235</th>\n",
       "      <td>Corynebacterium diphtheriae</td>\n",
       "      <td>NCTC13129</td>\n",
       "      <td>http://www.wikidata.org/entity/Q131649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1591</th>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>10158/10184</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8240</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus Mu50</td>\n",
       "      <td>Mu50</td>\n",
       "      <td>http://www.wikidata.org/entity/Q188121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4597</th>\n",
       "      <td>Burkholderia multivorans</td>\n",
       "      <td>AU23919</td>\n",
       "      <td>http://www.wikidata.org/entity/Q4999018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7471</th>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "      <td></td>\n",
       "      <td>http://www.wikidata.org/entity/Q221179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     _organism      _strain  \\\n",
       "7235               Corynebacterium diphtheriae    NCTC13129   \n",
       "1591                          Escherichia coli  10158/10184   \n",
       "8240  Staphylococcus aureus subsp. aureus Mu50         Mu50   \n",
       "4597                  Burkholderia multivorans      AU23919   \n",
       "7471                  Streptococcus pneumoniae                \n",
       "\n",
       "                                 parent taxon  \n",
       "7235   http://www.wikidata.org/entity/Q131649  \n",
       "1591                                     None  \n",
       "8240   http://www.wikidata.org/entity/Q188121  \n",
       "4597  http://www.wikidata.org/entity/Q4999018  \n",
       "7471   http://www.wikidata.org/entity/Q221179  "
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "def search_parent_taxon(df_row: pd.Series) -> str:\n",
    "    \"\"\"\n",
    "    Searches in wikidata for a species of bacterium which matches the two first words of _organism\n",
    "    Sometime we have abbreviations for example Achromobacter sp. -- nobody knows if this is rather Achromobacter spanius or Achromobacter spiritinus -- This is all rather unusable\n",
    "    \"\"\"\n",
    "    parent = df_row._organism.split()[:2]\n",
    "    abbreviation = False\n",
    "    for i, word in enumerate(parent): \n",
    "        if word[-1] == \".\": \n",
    "            abbreviation = True\n",
    "            parent[i] = word[:-1]\n",
    "    parent = \" \".join(parent).lower()\n",
    "    query = f\"\"\"SELECT ?item ?itemLabel ?itemDescription\n",
    "    WHERE {{\n",
    "      ?item rdfs:label ?label;\n",
    "            schema:description \"species of bacterium\"@en.\n",
    "      \n",
    "      FILTER(LANG(?label) = \"en\" && CONTAINS(LCASE(?label), \"{parent}\"))\n",
    "      \n",
    "      SERVICE wikibase:label {{ bd:serviceParam wikibase:language \"[AUTO_LANGUAGE],en\". }}\n",
    "    }}\n",
    "    LIMIT 10\n",
    "    \"\"\"\n",
    "    sparql.setQuery(query)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    results = sparql.query().convert().get(\"results\").get(\"bindings\")\n",
    "    results = [(item.get(\"item\").get(\"value\"), item.get(\"itemLabel\").get(\"value\")) for item in results]\n",
    "    if abbreviation: \n",
    "        return [res[0] for res in results] if len(results) > 1 else results[0][0]\n",
    "    else: \n",
    "        for res in results: \n",
    "            if res[1].lower() in parent: \n",
    "                return res[0]\n",
    "\n",
    "dummy[\"parent taxon\"] = dummy.apply(search_parent_taxon, axis=1)\n",
    "dummy[[\"_organism\", \"_strain\", \"parent taxon\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "e7ec62ba-c5e6-4ca8-8f88-f447a91eb7fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'http://www.wikidata.org/entity/Q188121'"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dummy[\"parent taxon\"].iat[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad87fbec-626d-40a1-93f9-25ff9994cfc6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a65eb325-6ed8-4182-a68e-9b9a257734a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c493921-885c-43f1-b408-88c9729c6126",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd766a73-bef8-4965-95aa-a9637101c886",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48176951-a2f2-49ad-97cf-531e2111a95a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af7b3f22-bdb2-4a07-adab-a7791d5786f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop rows if _strain is empty and _organism has no numeric value in it -- expert knowledge required -- database shit \n",
    "\n",
    "dummy = dummy[np.logical_or(dummy[\"_strain\"].astype(bool), dummy[\"_organism\"].apply(lambda x: any(char.isdigit() for char in x)))]\n",
    "dummy"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lodkg",
   "language": "python",
   "name": "lodkg"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
